// @mysten/deepbook-v3/dist/pyth/pyth.mjs
import { PriceServiceConnection } from "./PriceServiceConnection.mjs";
import { extractVaaBytesFromAccumulatorMessage } from "./pyth-helpers.mjs";
import { State } from "../contracts/pyth/state.mjs";
import { State as State$1 } from "../contracts/wormhole/state.mjs";
import { bcs } from "@mysten/sui/bcs";
import { coinWithBalance } from "@mysten/sui/transactions";
import { fromBase64, fromHex, parseStructTag } from "@mysten/sui/utils";

//#region src/pyth/pyth.ts
const MAX_ARGUMENT_SIZE = 16 * 1024;
var SuiPriceServiceConnection = class extends PriceServiceConnection {
	/**
	* Fetch price feed update data.
	*
	* @param priceIds Array of hex-encoded price IDs.
	* @returns Array of buffers containing the price update data.
	*/
	async getPriceFeedsUpdateData(priceIds) {
		return (await this.getLatestVaas(priceIds)).map((vaa) => fromBase64(vaa));
	}
};
var SuiPythClient = class {
	#pythPackageId;
	#wormholePackageId;
	#priceFeedObjectIdCache = /* @__PURE__ */ new Map();
	#priceTableInfo;
	#baseUpdateFee;
	constructor(provider, pythStateId, wormholeStateId) {
		this.provider = provider;
		this.pythStateId = pythStateId;
		this.wormholeStateId = wormholeStateId;
	}
	/**
	* Verifies the VAAs using the Wormhole contract.
	*
	* @param vaas Array of VAA buffers to verify.
	* @param tx Transaction block to add commands to.
	* @returns Array of verified VAAs.
	*/
	async verifyVaas(vaas, tx) {
		const wormholePackageId = await this.getWormholePackageId();
		const verifiedVaas = [];
		for (const vaa of vaas) {
			const [verifiedVaa] = tx.moveCall({
				target: `${wormholePackageId}::vaa::parse_and_verify`,
				arguments: [
					tx.object(this.wormholeStateId),
					tx.pure.vector("u8", vaa),
					tx.object.clock()
				]
			});
			verifiedVaas.push(verifiedVaa);
		}
		return verifiedVaas;
	}
	/**
	* Adds the necessary commands for updating the Pyth price feeds to the transaction block.
	*
	* @param tx Transaction block to add commands to.
	* @param updates Array of price feed updates received from the price service.
	* @param feedIds Array of feed IDs to update (in hex format).
	*/
	async updatePriceFeeds(tx, updates, feedIds) {
		const packageId = await this.getPythPackageId();
		let priceUpdatesHotPotato;
		if (updates.length > 1) throw new Error("SDK does not support sending multiple accumulator messages in a single transaction");
		const vaa = extractVaaBytesFromAccumulatorMessage(updates[0]);
		const verifiedVaas = await this.verifyVaas([vaa], tx);
		[priceUpdatesHotPotato] = tx.moveCall({
			target: `${packageId}::pyth::create_authenticated_price_infos_using_accumulator`,
			arguments: [
				tx.object(this.pythStateId),
				tx.pure(bcs.vector(bcs.U8).serialize(Array.from(updates[0]), { maxSize: MAX_ARGUMENT_SIZE }).toBytes()),
				verifiedVaas[0],
				tx.object.clock()
			]
		});
		const priceInfoObjects = [];
		const baseUpdateFee = await this.getBaseUpdateFee();
		for (const feedId of feedIds) {
			const priceInfoObjectId = await this.getPriceFeedObjectId(feedId);
			if (!priceInfoObjectId) throw new Error(`Price feed ${feedId} not found, please create it first`);
			priceInfoObjects.push(priceInfoObjectId);
			[priceUpdatesHotPotato] = tx.moveCall({
				target: `${packageId}::pyth::update_single_price_feed`,
				arguments: [
					tx.object(this.pythStateId),
					priceUpdatesHotPotato,
					tx.object(priceInfoObjectId),
					coinWithBalance({ balance: baseUpdateFee }),
					tx.object.clock()
				]
			});
		}
		tx.moveCall({
			target: `${packageId}::hot_potato_vector::destroy`,
			arguments: [priceUpdatesHotPotato],
			typeArguments: [`${packageId}::price_info::PriceInfo`]
		});
		return priceInfoObjects;
	}
	/**
	* Get the price feed object ID for a given feed ID, caching the promise.
	* @param feedId
	*/
	getPriceFeedObjectId(feedId) {
		if (!this.#priceFeedObjectIdCache.has(feedId)) this.#priceFeedObjectIdCache.set(feedId, this.#fetchPriceFeedObjectId(feedId).catch((err) => {
			this.#priceFeedObjectIdCache.delete(feedId);
			throw err;
		}));
		return this.#priceFeedObjectIdCache.get(feedId);
	}
	/**
	* Fetches the price feed object ID for a given feed ID (no caching).
	* Throws an error if the object is not found.
	*/
	async #fetchPriceFeedObjectId(feedId) {
		const { id: tableId, fieldType } = await this.getPriceTableInfo();
		const nameBytes = bcs.byteVector().serialize(Array.from(fromHex(feedId))).toBytes();
		const result = await this.provider.core.getDynamicField({
			parentId: tableId,
			name: {
				type: `${fieldType}::price_identifier::PriceIdentifier`,
				bcs: nameBytes
			}
		});
		if (!result.dynamicField) throw new Error(`Price feed object ID for feed ID ${feedId} not found.`);
		return bcs.Address.parse(result.dynamicField.value.bcs);
	}
	/**
	* Fetches the price table object ID for the current state ID, caching the promise.
	* @returns Price table object ID and field type
	*/
	getPriceTableInfo() {
		if (!this.#priceTableInfo) this.#priceTableInfo = this.#fetchPriceTableInfo().catch((err) => {
			this.#priceTableInfo = void 0;
			throw err;
		});
		return this.#priceTableInfo;
	}
	/**
	* Fetches the price table object ID and field type (no caching).
	* @returns Price table object ID and field type
	*/
	async #fetchPriceTableInfo() {
		const result = await this.provider.core.getDynamicObjectField({
			parentId: this.pythStateId,
			name: {
				type: "vector<u8>",
				bcs: bcs.string().serialize("price_info").toBytes()
			}
		});
		if (!result.object) throw new Error("Price Table not found, contract may not be initialized");
		const priceIdentifier = parseStructTag(result.object.type).typeParams[0];
		if (typeof priceIdentifier === "object" && priceIdentifier !== null && priceIdentifier.name === "PriceIdentifier" && "address" in priceIdentifier) return {
			id: result.object.objectId,
			fieldType: priceIdentifier.address
		};
		else throw new Error("fieldType not found");
	}
	/**
	* Fetches the package ID for the Wormhole contract, with caching.
	*/
	getWormholePackageId() {
		if (!this.#wormholePackageId) this.#wormholePackageId = this.#fetchWormholePackageId();
		return this.#wormholePackageId;
	}
	/**
	* Fetches the package ID for the Wormhole contract (no caching).
	*/
	async #fetchWormholePackageId() {
		const result = await this.provider.core.getObject({
			objectId: this.wormholeStateId,
			include: { content: true }
		});
		if (!result.object?.content) throw new Error("Unable to fetch Wormhole state object");
		return State$1.parse(result.object.content).upgrade_cap.package;
	}
	/**
	* Fetches the package ID for the Pyth contract, with caching.
	*/
	getPythPackageId() {
		if (!this.#pythPackageId) this.#pythPackageId = this.#fetchPythPackageId();
		return this.#pythPackageId;
	}
	/**
	* Fetches the package ID for the Pyth contract (no caching).
	*/
	async #fetchPythPackageId() {
		const result = await this.provider.core.getObject({
			objectId: this.pythStateId,
			include: { content: true }
		});
		if (!result.object?.content) throw new Error(`Cannot fetch Pyth package ID for object ${this.pythStateId}`);
		return State.parse(result.object.content).upgrade_cap.package;
	}
	/**
	* Gets the base update fee from the Pyth state object.
	*/
	async #fetchBaseUpdateFee() {
		const result = await this.provider.core.getObject({
			objectId: this.pythStateId,
			include: { content: true }
		});
		if (!result.object?.content) throw new Error("Unable to fetch Pyth state object");
		const state = State.parse(result.object.content);
		return Number(state.base_update_fee);
	}
	/**
	* Returns the cached base update fee, fetching it if necessary.
	*/
	getBaseUpdateFee() {
		if (!this.#baseUpdateFee) this.#baseUpdateFee = this.#fetchBaseUpdateFee();
		return this.#baseUpdateFee;
	}
};

//#endregion
export { SuiPriceServiceConnection, SuiPythClient };
