// @mysten/deepbook-v3/dist/queries/balanceManagerQueries.mjs
import { bcs } from "@mysten/sui/bcs";
import { Transaction } from "@mysten/sui/transactions";
import { normalizeSuiAddress } from "@mysten/sui/utils";

//#region src/queries/balanceManagerQueries.ts
var BalanceManagerQueries = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	async checkManagerBalance(managerKey, coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		const coin = this.#ctx.config.getCoin(coinKey);
		tx.add(this.#ctx.balanceManager.checkManagerBalance(managerKey, coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		const parsed_balance = bcs.U64.parse(bytes);
		const adjusted_balance = Number(parsed_balance) / coin.scalar;
		return {
			coinType: coin.type,
			balance: Number(adjusted_balance.toFixed(9))
		};
	}
	async checkManagerBalanceWithAddress(managerAddress, coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		const coin = this.#ctx.config.getCoin(coinKey);
		tx.moveCall({
			target: `${this.#ctx.config.DEEPBOOK_PACKAGE_ID}::balance_manager::balance`,
			arguments: [tx.object(managerAddress)],
			typeArguments: [coin.type]
		});
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		const parsed_balance = bcs.U64.parse(bytes);
		const adjusted_balance = Number(parsed_balance) / coin.scalar;
		return {
			coinType: coin.type,
			balance: Number(adjusted_balance.toFixed(9))
		};
	}
	async checkManagerBalancesWithAddress(managerAddresses, coinKeys) {
		if (managerAddresses.length === 0 || coinKeys.length === 0) return {};
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		const coins = coinKeys.map((coinKey) => this.#ctx.config.getCoin(coinKey));
		for (const managerAddress of managerAddresses) for (const coin of coins) tx.moveCall({
			target: `${this.#ctx.config.DEEPBOOK_PACKAGE_ID}::balance_manager::balance`,
			arguments: [tx.object(managerAddress)],
			typeArguments: [coin.type]
		});
		const res = await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		});
		if (res.FailedTransaction) throw new Error(`Transaction failed: ${res.FailedTransaction.status.error?.message || "Unknown error"}`);
		if (!res.commandResults) throw new Error("Failed to get manager balances: No command results");
		const results = {};
		for (let m = 0; m < managerAddresses.length; m++) {
			const managerAddress = managerAddresses[m];
			const managerBalances = {};
			for (let c = 0; c < coins.length; c++) {
				const coin = coins[c];
				const commandResult = res.commandResults[m * coins.length + c];
				if (!commandResult || !commandResult.returnValues) throw new Error(`Failed to get balance for ${coin.type}: No return values`);
				const bytes = commandResult.returnValues[0].bcs;
				const parsed_balance = bcs.U64.parse(bytes);
				managerBalances[coin.type] = Number((Number(parsed_balance) / coin.scalar).toFixed(9));
			}
			results[managerAddress] = managerBalances;
		}
		return results;
	}
	async getBalanceManagerIds(owner) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.deepBook.getBalanceManagerIds(owner));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return bcs.vector(bcs.Address).parse(bytes).map((id) => normalizeSuiAddress(id));
	}
	async accountExists(poolKey, managerKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.deepBook.accountExists(poolKey, managerKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return bcs.bool().parse(bytes);
	}
};

//#endregion
export { BalanceManagerQueries };
