// @mysten/deepbook-v3/dist/queries/marginPoolQueries.mjs
import { FLOAT_SCALAR } from "../utils/config.mjs";
import { formatTokenAmount } from "./context.mjs";
import { bcs } from "@mysten/sui/bcs";
import { Transaction } from "@mysten/sui/transactions";

//#region src/queries/marginPoolQueries.ts
var MarginPoolQueries = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	async getMarginPoolId(coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.getId(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return bcs.Address.parse(bytes);
	}
	async isDeepbookPoolAllowed(coinKey, deepbookPoolId) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.deepbookPoolAllowed(coinKey, deepbookPoolId));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return bcs.bool().parse(bytes);
	}
	async getMarginPoolTotalSupply(coinKey, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.totalSupply(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getMarginPoolSupplyShares(coinKey, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.supplyShares(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getMarginPoolTotalBorrow(coinKey, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.totalBorrow(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getMarginPoolBorrowShares(coinKey, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.borrowShares(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getMarginPoolLastUpdateTimestamp(coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.lastUpdateTimestamp(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes));
	}
	async getMarginPoolSupplyCap(coinKey, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.supplyCap(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getMarginPoolMaxUtilizationRate(coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.maxUtilizationRate(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getMarginPoolProtocolSpread(coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.protocolSpread(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getMarginPoolMinBorrow(coinKey, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.minBorrow(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getMarginPoolInterestRate(coinKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.interestRate(coinKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getUserSupplyShares(coinKey, supplierCapId, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.userSupplyShares(coinKey, supplierCapId));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
	async getUserSupplyAmount(coinKey, supplierCapId, decimals = 6) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginPool.userSupplyAmount(coinKey, supplierCapId));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return formatTokenAmount(BigInt(bcs.U64.parse(bytes)), this.#ctx.config.getCoin(coinKey).scalar, decimals);
	}
};

//#endregion
export { MarginPoolQueries };
