// @mysten/deepbook-v3/dist/queries/priceFeedQueries.mjs
import { PRICE_INFO_OBJECT_MAX_AGE_MS } from "../utils/config.mjs";
import { PriceInfoObject } from "../contracts/pyth/price_info.mjs";
import { SuiPriceServiceConnection, SuiPythClient } from "../pyth/pyth.mjs";

//#region src/queries/priceFeedQueries.ts
var PriceFeedQueries = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	async getPriceInfoObject(tx, coinKey) {
		this.#ctx.config.requirePyth();
		const currentTime = Date.now();
		const priceInfoObjectAge = await this.getPriceInfoObjectAge(coinKey);
		if (priceInfoObjectAge && currentTime - priceInfoObjectAge * 1e3 < PRICE_INFO_OBJECT_MAX_AGE_MS) return await this.#ctx.config.getCoin(coinKey).priceInfoObjectId;
		const connection = new SuiPriceServiceConnection(this.#ctx.config.network === "testnet" ? "https://hermes-beta.pyth.network" : "https://hermes.pyth.network");
		const priceIDs = [this.#ctx.config.getCoin(coinKey).feed];
		const priceUpdateData = await connection.getPriceFeedsUpdateData(priceIDs);
		const wormholeStateId = this.#ctx.config.pyth.wormholeStateId;
		const pythStateId = this.#ctx.config.pyth.pythStateId;
		return (await new SuiPythClient(this.#ctx.client, pythStateId, wormholeStateId).updatePriceFeeds(tx, priceUpdateData, priceIDs))[0];
	}
	async getPriceInfoObjects(tx, coinKeys) {
		this.#ctx.config.requirePyth();
		if (coinKeys.length === 0) return {};
		const currentTime = Date.now();
		const coinToObjectId = {};
		const objectIds = [];
		for (const coinKey of coinKeys) {
			const priceInfoObjectId = this.#ctx.config.getCoin(coinKey).priceInfoObjectId;
			coinToObjectId[coinKey] = priceInfoObjectId;
			objectIds.push(priceInfoObjectId);
		}
		const res = await this.#ctx.client.core.getObjects({
			objectIds,
			include: { content: true }
		});
		const staleCoinKeys = [];
		const result = {};
		for (let i = 0; i < coinKeys.length; i++) {
			const coinKey = coinKeys[i];
			const obj = res.objects[i];
			if (obj instanceof Error || !obj?.content) {
				staleCoinKeys.push(coinKey);
				continue;
			}
			const priceInfoObject = PriceInfoObject.parse(obj.content);
			if (currentTime - Number(priceInfoObject.price_info.arrival_time) * 1e3 >= PRICE_INFO_OBJECT_MAX_AGE_MS) staleCoinKeys.push(coinKey);
			else result[coinKey] = coinToObjectId[coinKey];
		}
		if (staleCoinKeys.length === 0) return result;
		const staleFeedIds = [];
		const feedIdToCoinKey = {};
		for (const coinKey of staleCoinKeys) {
			const feedId = this.#ctx.config.getCoin(coinKey).feed;
			staleFeedIds.push(feedId);
			feedIdToCoinKey[feedId] = coinKey;
		}
		const priceUpdateData = await new SuiPriceServiceConnection(this.#ctx.config.network === "testnet" ? "https://hermes-beta.pyth.network" : "https://hermes.pyth.network").getPriceFeedsUpdateData(staleFeedIds);
		const wormholeStateId = this.#ctx.config.pyth.wormholeStateId;
		const pythStateId = this.#ctx.config.pyth.pythStateId;
		const updatedObjectIds = await new SuiPythClient(this.#ctx.client, pythStateId, wormholeStateId).updatePriceFeeds(tx, priceUpdateData, staleFeedIds);
		for (let i = 0; i < staleFeedIds.length; i++) {
			const coinKey = feedIdToCoinKey[staleFeedIds[i]];
			result[coinKey] = updatedObjectIds[i];
		}
		return result;
	}
	async getPriceInfoObjectAge(coinKey) {
		const priceInfoObjectId = this.#ctx.config.getCoin(coinKey).priceInfoObjectId;
		const res = await this.#ctx.client.core.getObject({
			objectId: priceInfoObjectId,
			include: { content: true }
		});
		if (!res.object?.content) throw new Error(`Price info object not found for ${coinKey}`);
		const priceInfoObject = PriceInfoObject.parse(res.object.content);
		return Number(priceInfoObject.price_info.arrival_time);
	}
};

//#endregion
export { PriceFeedQueries };
