// @mysten/deepbook-v3/dist/queries/registryQueries.mjs
import { VecSet } from "../contracts/deepbook/deps/sui/vec_set.mjs";
import "../types/bcs.mjs";
import { FLOAT_SCALAR } from "../utils/config.mjs";
import { bcs } from "@mysten/sui/bcs";
import { Transaction } from "@mysten/sui/transactions";
import { normalizeSuiAddress } from "@mysten/sui/utils";

//#region src/queries/registryQueries.ts
var RegistryQueries = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	async isPoolEnabledForMargin(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.poolEnabled(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return bcs.Bool.parse(bytes);
	}
	async getMarginManagerIdsForOwner(owner) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.getMarginManagerIds(owner));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return VecSet(bcs.Address).parse(bytes).contents.map((id) => normalizeSuiAddress(id));
	}
	async getBaseMarginPoolId(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.baseMarginPoolId(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return "0x" + bcs.Address.parse(bytes);
	}
	async getQuoteMarginPoolId(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.quoteMarginPoolId(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return "0x" + bcs.Address.parse(bytes);
	}
	async getMinWithdrawRiskRatio(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.minWithdrawRiskRatio(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getMinBorrowRiskRatio(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.minBorrowRiskRatio(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getLiquidationRiskRatio(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.liquidationRiskRatio(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getTargetLiquidationRiskRatio(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.targetLiquidationRiskRatio(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getUserLiquidationReward(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.userLiquidationReward(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getPoolLiquidationReward(poolKey) {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.poolLiquidationReward(poolKey));
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return Number(bcs.U64.parse(bytes)) / FLOAT_SCALAR;
	}
	async getAllowedMaintainers() {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.allowedMaintainers());
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return VecSet(bcs.Address).parse(bytes).contents.map((id) => normalizeSuiAddress(id));
	}
	async getAllowedPauseCaps() {
		const tx = new Transaction();
		tx.setSender(this.#ctx.address);
		tx.add(this.#ctx.marginRegistry.allowedPauseCaps());
		const bytes = (await this.#ctx.client.core.simulateTransaction({
			transaction: tx,
			include: {
				commandResults: true,
				effects: true
			}
		})).commandResults[0].returnValues[0].bcs;
		return VecSet(bcs.Address).parse(bytes).contents.map((id) => normalizeSuiAddress(id));
	}
};

//#endregion
export { RegistryQueries };
