// @mysten/deepbook-v3/dist/contracts/wormhole/state.mjs
import { MoveStruct } from "../utils/index.mjs";
import { ExternalAddress } from "./external_address.mjs";
import { Table } from "./deps/0x0000000000000000000000000000000000000000000000000000000000000002/table.mjs";
import { ConsumedVAAs } from "./consumed_vaas.mjs";
import { FeeCollector } from "./fee_collector.mjs";
import { UpgradeCap } from "./deps/0x0000000000000000000000000000000000000000000000000000000000000002/package.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/wormhole/state.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "0xf47329f4344f3bf0f8e436e2f7b485466cff300f12a166563995d3888c296a94::state";
const State = new MoveStruct({
	name: `${$moduleName}::State`,
	fields: {
		id: bcs.Address,
		governance_chain: bcs.u16(),
		governance_contract: ExternalAddress,
		guardian_set_index: bcs.u32(),
		guardian_sets: Table,
		guardian_set_seconds_to_live: bcs.u32(),
		consumed_vaas: ConsumedVAAs,
		fee_collector: FeeCollector,
		upgrade_cap: UpgradeCap
	}
});

//#endregion
export { State };
