// @mysten/sui/dist/grpc/proto/sui/forking/v1alpha/forking_service.client.mjs
import { ForkingService } from "./forking_service.mjs";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";

//#region src/grpc/proto/sui/forking/v1alpha/forking_service.client.ts
/**
* Administrative service for controlling a forked Sui network.
*
* @generated from protobuf service sui.forking.v1alpha.ForkingService
*/
var ForkingServiceClient = class {
	constructor(_transport) {
		this._transport = _transport;
		this.typeName = ForkingService.typeName;
		this.methods = ForkingService.methods;
		this.options = ForkingService.options;
	}
	/**
	* Advance the forked network's clock by a given duration.
	*
	* @generated from protobuf rpc: AdvanceClock(sui.forking.v1alpha.AdvanceClockRequest) returns (sui.forking.v1alpha.AdvanceClockResponse);
	*/
	advanceClock(input, options) {
		const method = this.methods[0], opt = this._transport.mergeOptions(options);
		return stackIntercept("unary", this._transport, method, opt, input);
	}
	/**
	* Seal pending transactions into a new checkpoint.
	*
	* @generated from protobuf rpc: AdvanceCheckpoint(sui.forking.v1alpha.AdvanceCheckpointRequest) returns (sui.forking.v1alpha.AdvanceCheckpointResponse);
	*/
	advanceCheckpoint(input, options) {
		const method = this.methods[1], opt = this._transport.mergeOptions(options);
		return stackIntercept("unary", this._transport, method, opt, input);
	}
	/**
	* Return the current state of the forked network.
	*
	* @generated from protobuf rpc: GetStatus(sui.forking.v1alpha.GetStatusRequest) returns (sui.forking.v1alpha.GetStatusResponse);
	*/
	getStatus(input, options) {
		const method = this.methods[2], opt = this._transport.mergeOptions(options);
		return stackIntercept("unary", this._transport, method, opt, input);
	}
};

//#endregion
export { ForkingServiceClient };
