// @mysten/sui/dist/grpc/proto/sui/forking/v1alpha/forking_service.mjs
import { MessageType } from "@protobuf-ts/runtime";
import { ServiceType } from "@protobuf-ts/runtime-rpc";

//#region src/grpc/proto/sui/forking/v1alpha/forking_service.ts
var AdvanceClockRequest$Type = class extends MessageType {
	constructor() {
		super("sui.forking.v1alpha.AdvanceClockRequest", [{
			no: 1,
			name: "duration_ms",
			kind: "scalar",
			opt: true,
			T: 4,
			L: 0
		}]);
	}
};
/**
* @generated MessageType for protobuf message sui.forking.v1alpha.AdvanceClockRequest
*/
const AdvanceClockRequest = new AdvanceClockRequest$Type();
var AdvanceClockResponse$Type = class extends MessageType {
	constructor() {
		super("sui.forking.v1alpha.AdvanceClockResponse", [{
			no: 1,
			name: "timestamp_ms",
			kind: "scalar",
			T: 4,
			L: 0
		}, {
			no: 2,
			name: "tx_digest",
			kind: "scalar",
			T: 9
		}]);
	}
};
/**
* @generated MessageType for protobuf message sui.forking.v1alpha.AdvanceClockResponse
*/
const AdvanceClockResponse = new AdvanceClockResponse$Type();
var AdvanceCheckpointRequest$Type = class extends MessageType {
	constructor() {
		super("sui.forking.v1alpha.AdvanceCheckpointRequest", []);
	}
};
/**
* @generated MessageType for protobuf message sui.forking.v1alpha.AdvanceCheckpointRequest
*/
const AdvanceCheckpointRequest = new AdvanceCheckpointRequest$Type();
var AdvanceCheckpointResponse$Type = class extends MessageType {
	constructor() {
		super("sui.forking.v1alpha.AdvanceCheckpointResponse", [{
			no: 1,
			name: "checkpoint_sequence_number",
			kind: "scalar",
			T: 4,
			L: 0
		}, {
			no: 2,
			name: "timestamp_ms",
			kind: "scalar",
			T: 4,
			L: 0
		}]);
	}
};
/**
* @generated MessageType for protobuf message sui.forking.v1alpha.AdvanceCheckpointResponse
*/
const AdvanceCheckpointResponse = new AdvanceCheckpointResponse$Type();
var GetStatusRequest$Type = class extends MessageType {
	constructor() {
		super("sui.forking.v1alpha.GetStatusRequest", []);
	}
};
/**
* @generated MessageType for protobuf message sui.forking.v1alpha.GetStatusRequest
*/
const GetStatusRequest = new GetStatusRequest$Type();
var GetStatusResponse$Type = class extends MessageType {
	constructor() {
		super("sui.forking.v1alpha.GetStatusResponse", [
			{
				no: 1,
				name: "epoch",
				kind: "scalar",
				T: 4,
				L: 0
			},
			{
				no: 2,
				name: "checkpoint_sequence_number",
				kind: "scalar",
				T: 4,
				L: 0
			},
			{
				no: 3,
				name: "timestamp_ms",
				kind: "scalar",
				T: 4,
				L: 0
			},
			{
				no: 4,
				name: "forked_at_checkpoint",
				kind: "scalar",
				T: 4,
				L: 0
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.forking.v1alpha.GetStatusResponse
*/
const GetStatusResponse = new GetStatusResponse$Type();
/**
* @generated ServiceType for protobuf service sui.forking.v1alpha.ForkingService
*/
const ForkingService = new ServiceType("sui.forking.v1alpha.ForkingService", [
	{
		name: "AdvanceClock",
		options: {},
		I: AdvanceClockRequest,
		O: AdvanceClockResponse
	},
	{
		name: "AdvanceCheckpoint",
		options: {},
		I: AdvanceCheckpointRequest,
		O: AdvanceCheckpointResponse
	},
	{
		name: "GetStatus",
		options: {},
		I: GetStatusRequest,
		O: GetStatusResponse
	}
]);

//#endregion
export { AdvanceCheckpointRequest, AdvanceCheckpointResponse, AdvanceClockRequest, AdvanceClockResponse, ForkingService, GetStatusRequest, GetStatusResponse };
