// @mysten/deepbook-v3/dist/utils/config.mjs
import { BalanceManagerContract } from "../transactions/balanceManager.mjs";
import { ConfigurationError, ErrorMessages, ResourceNotFoundError } from "./errors.mjs";
import { mainnetCoins, mainnetMarginPools, mainnetPackageIds, mainnetPools, mainnetPythConfigs, testnetCoins, testnetMarginPools, testnetPackageIds, testnetPools, testnetPythConfigs } from "./constants.mjs";
import { normalizeSuiAddress } from "@mysten/sui/utils";

//#region src/utils/config.ts
const FLOAT_SCALAR = 1e9;
const DEEP_SCALAR = 1e6;
const MAX_TIMESTAMP = 1844674407370955161n;
const PRICE_INFO_OBJECT_MAX_AGE_MS = 3e4;
const GAS_BUDGET = 25e7;
const POOL_CREATION_FEE_DEEP = 5e8;
var DeepBookConfig = class {
	#coins;
	#pools;
	#marginPools;
	constructor({ network, address, adminCap, marginAdminCap, marginMaintainerCap, balanceManagers, marginManagers, coins, pools, marginPools, packageIds, pyth }) {
		this.network = network;
		this.address = normalizeSuiAddress(address);
		this.adminCap = adminCap;
		this.marginAdminCap = marginAdminCap;
		this.marginMaintainerCap = marginMaintainerCap;
		this.balanceManagers = balanceManagers || {};
		this.marginManagers = marginManagers || {};
		if (packageIds) {
			this.DEEPBOOK_PACKAGE_ID = packageIds.DEEPBOOK_PACKAGE_ID || "";
			this.REGISTRY_ID = packageIds.REGISTRY_ID || "";
			this.DEEP_TREASURY_ID = packageIds.DEEP_TREASURY_ID || "";
			this.MARGIN_PACKAGE_ID = packageIds.MARGIN_PACKAGE_ID || "";
			this.MARGIN_V1 = packageIds.MARGIN_V1 || "";
			this.MARGIN_REGISTRY_ID = packageIds.MARGIN_REGISTRY_ID || "";
			this.LIQUIDATION_PACKAGE_ID = packageIds.LIQUIDATION_PACKAGE_ID || "";
			this.#coins = coins || {};
			this.#pools = pools || {};
			this.#marginPools = marginPools || {};
			this.pyth = pyth || {
				pythStateId: "",
				wormholeStateId: ""
			};
		} else if (network === "mainnet") {
			this.#coins = coins || mainnetCoins;
			this.#pools = pools || mainnetPools;
			this.#marginPools = marginPools || mainnetMarginPools;
			this.DEEPBOOK_PACKAGE_ID = mainnetPackageIds.DEEPBOOK_PACKAGE_ID;
			this.REGISTRY_ID = mainnetPackageIds.REGISTRY_ID;
			this.DEEP_TREASURY_ID = mainnetPackageIds.DEEP_TREASURY_ID;
			this.MARGIN_PACKAGE_ID = mainnetPackageIds.MARGIN_PACKAGE_ID;
			this.MARGIN_V1 = mainnetPackageIds.MARGIN_V1;
			this.MARGIN_REGISTRY_ID = mainnetPackageIds.MARGIN_REGISTRY_ID;
			this.LIQUIDATION_PACKAGE_ID = mainnetPackageIds.LIQUIDATION_PACKAGE_ID;
			this.pyth = mainnetPythConfigs;
		} else if (network === "testnet") {
			this.#coins = coins || testnetCoins;
			this.#pools = pools || testnetPools;
			this.#marginPools = marginPools || testnetMarginPools;
			this.DEEPBOOK_PACKAGE_ID = testnetPackageIds.DEEPBOOK_PACKAGE_ID;
			this.REGISTRY_ID = testnetPackageIds.REGISTRY_ID;
			this.DEEP_TREASURY_ID = testnetPackageIds.DEEP_TREASURY_ID;
			this.MARGIN_PACKAGE_ID = testnetPackageIds.MARGIN_PACKAGE_ID;
			this.MARGIN_V1 = testnetPackageIds.MARGIN_V1;
			this.MARGIN_REGISTRY_ID = testnetPackageIds.MARGIN_REGISTRY_ID;
			this.LIQUIDATION_PACKAGE_ID = testnetPackageIds.LIQUIDATION_PACKAGE_ID;
			this.pyth = testnetPythConfigs;
		} else throw new Error(`Network '${network}' is not supported by default. Provide custom 'packageIds' for non-standard networks.`);
		this.balanceManager = new BalanceManagerContract(this);
	}
	requirePyth() {
		if (!this.pyth.pythStateId || !this.pyth.wormholeStateId) throw new ConfigurationError("Pyth configuration is required for price feed operations. Provide 'pyth' when using custom packageIds.");
	}
	getCoin(key) {
		const coin = this.#coins[key];
		if (!coin) throw new ResourceNotFoundError("Coin", key);
		return coin;
	}
	getPool(key) {
		const pool = this.#pools[key];
		if (!pool) throw new ResourceNotFoundError("Pool", key);
		return pool;
	}
	getMarginPool(key) {
		const pool = this.#marginPools[key];
		if (!pool) throw new ResourceNotFoundError("Margin pool", key);
		return pool;
	}
	/**
	* @description Get the balance manager by key
	* @param managerKey Key of the balance manager
	* @returns The BalanceManager object
	*/
	getBalanceManager(managerKey) {
		if (!Object.hasOwn(this.balanceManagers, managerKey)) throw new Error(ErrorMessages.BALANCE_MANAGER_NOT_FOUND(managerKey));
		return this.balanceManagers[managerKey];
	}
	/**
	* @description Get the margin manager by key
	* @param managerKey Key of the margin manager
	* @returns The MarginManager object
	*/
	getMarginManager(managerKey) {
		if (!Object.hasOwn(this.marginManagers, managerKey)) throw new Error(ErrorMessages.MARGIN_MANAGER_NOT_FOUND(managerKey));
		return this.marginManagers[managerKey];
	}
};

//#endregion
export { DEEP_SCALAR, DeepBookConfig, FLOAT_SCALAR, GAS_BUDGET, MAX_TIMESTAMP, POOL_CREATION_FEE_DEEP, PRICE_INFO_OBJECT_MAX_AGE_MS };
