// @mysten/deepbook-v3/dist/utils/validation.mjs
import { ConfigurationError, ValidationError } from "./errors.mjs";
import { isValidSuiAddress } from "@mysten/sui/utils";

//#region src/utils/validation.ts
/**
* @description Validate that a required configuration value is set
* @param {T | undefined} value - The value to validate
* @param {string} errorMessage - Error message if validation fails
* @returns {T} The validated value
* @throws {ConfigurationError} If the value is not set
*/
function validateRequired(value, errorMessage) {
	if (value === void 0 || value === null) throw new ConfigurationError(errorMessage);
	return value;
}
/**
* @description Validate that a Sui address is valid
* @param {string} address - The address to validate
* @param {string} [fieldName] - Name of the field for error messages (default 'Address')
* @returns {string} The validated address
* @throws {ValidationError} If the address is invalid
*/
function validateAddress(address, fieldName = "Address") {
	if (!isValidSuiAddress(address)) throw new ValidationError(`${fieldName} must be a valid Sui address`);
	return address;
}
/**
* @description Validate that a number is positive
* @param {number} value - The value to validate
* @param {string} fieldName - Name of the field for error messages
* @returns {number} The validated value
* @throws {ValidationError} If the value is not positive
*/
function validatePositiveNumber(value, fieldName) {
	if (value <= 0) throw new ValidationError(`${fieldName} must be a positive number`);
	return value;
}
/**
* @description Validate that a number is non-negative
* @param {number} value - The value to validate
* @param {string} fieldName - Name of the field for error messages
* @returns {number} The validated value
* @throws {ValidationError} If the value is negative
*/
function validateNonNegativeNumber(value, fieldName) {
	if (value < 0) throw new ValidationError(`${fieldName} must be non-negative`);
	return value;
}
/**
* @description Validate that a value is within a specified range
* @param {number} value - The value to validate
* @param {number} min - Minimum allowed value (inclusive)
* @param {number} max - Maximum allowed value (inclusive)
* @param {string} fieldName - Name of the field for error messages
* @returns {number} The validated value
* @throws {ValidationError} If the value is outside the range
*/
function validateRange(value, min, max, fieldName) {
	if (value < min || value > max) throw new ValidationError(`${fieldName} must be between ${min} and ${max}`);
	return value;
}
/**
* @description Validate array is not empty
* @param {T[]} array - The array to validate
* @param {string} fieldName - Name of the field for error messages
* @returns {T[]} The validated array
* @throws {ValidationError} If the array is empty
*/
function validateNonEmptyArray(array, fieldName) {
	if (!array || array.length === 0) throw new ValidationError(`${fieldName} cannot be empty`);
	return array;
}

//#endregion
export { validateAddress, validateNonEmptyArray, validateNonNegativeNumber, validatePositiveNumber, validateRange, validateRequired };
