// @mysten/deepbook-v3/dist/contracts/deepbook/deep_price.mjs
import { MoveStruct } from "../utils/index.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/deepbook/deep_price.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
/**
* DEEP price module. This module maintains the conversion rate between DEEP and
* the base and quote assets.
*/
const $moduleName = "@deepbook/core::deep_price";
const Price = new MoveStruct({
	name: `${$moduleName}::Price`,
	fields: {
		conversion_rate: bcs.u64(),
		timestamp: bcs.u64()
	}
});
const PriceAdded = new MoveStruct({
	name: `${$moduleName}::PriceAdded`,
	fields: {
		conversion_rate: bcs.u64(),
		timestamp: bcs.u64(),
		is_base_conversion: bcs.bool(),
		reference_pool: bcs.Address,
		target_pool: bcs.Address
	}
});
const DeepPrice = new MoveStruct({
	name: `${$moduleName}::DeepPrice`,
	fields: {
		base_prices: bcs.vector(Price),
		cumulative_base: bcs.u64(),
		quote_prices: bcs.vector(Price),
		cumulative_quote: bcs.u64()
	}
});
const OrderDeepPrice = new MoveStruct({
	name: `${$moduleName}::OrderDeepPrice`,
	fields: {
		asset_is_base: bcs.bool(),
		deep_per_asset: bcs.u64()
	}
});

//#endregion
export { OrderDeepPrice };
