// @mysten/deepbook-v3/dist/contracts/deepbook/order.mjs
import { MoveStruct } from "../utils/index.mjs";
import { OrderDeepPrice } from "./deep_price.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/deepbook/order.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
/**
* Order module defines the order struct and its methods. All order matching
* happens in this module.
*/
const $moduleName = "@deepbook/core::order";
const Order = new MoveStruct({
	name: `${$moduleName}::Order`,
	fields: {
		balance_manager_id: bcs.Address,
		order_id: bcs.u128(),
		client_order_id: bcs.u64(),
		quantity: bcs.u64(),
		filled_quantity: bcs.u64(),
		fee_is_deep: bcs.bool(),
		order_deep_price: OrderDeepPrice,
		epoch: bcs.u64(),
		status: bcs.u8(),
		expire_timestamp: bcs.u64()
	}
});
const OrderCanceled = new MoveStruct({
	name: `${$moduleName}::OrderCanceled`,
	fields: {
		balance_manager_id: bcs.Address,
		pool_id: bcs.Address,
		order_id: bcs.u128(),
		client_order_id: bcs.u64(),
		trader: bcs.Address,
		price: bcs.u64(),
		is_bid: bcs.bool(),
		original_quantity: bcs.u64(),
		base_asset_quantity_canceled: bcs.u64(),
		timestamp: bcs.u64()
	}
});
const OrderModified = new MoveStruct({
	name: `${$moduleName}::OrderModified`,
	fields: {
		balance_manager_id: bcs.Address,
		pool_id: bcs.Address,
		order_id: bcs.u128(),
		client_order_id: bcs.u64(),
		trader: bcs.Address,
		price: bcs.u64(),
		is_bid: bcs.bool(),
		previous_quantity: bcs.u64(),
		filled_quantity: bcs.u64(),
		new_quantity: bcs.u64(),
		timestamp: bcs.u64()
	}
});

//#endregion
export { Order };
