// @mysten/deepbook-v3/dist/contracts/pyth/state.mjs
import { MoveStruct } from "../utils/index.mjs";
import { DataSource } from "./data_source.mjs";
import { ConsumedVAAs } from "./deps/0xf47329f4344f3bf0f8e436e2f7b485466cff300f12a166563995d3888c296a94/consumed_vaas.mjs";
import { UpgradeCap } from "./deps/0x0000000000000000000000000000000000000000000000000000000000000002/package.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/pyth/state.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "0xabf837e98c26087cba0883c0a7a28326b1fa3c5e1e2c5abdb486f9e8f594c837::state";
const LatestOnly = new MoveStruct({
	name: `${$moduleName}::LatestOnly`,
	fields: { dummy_field: bcs.bool() }
});
const State = new MoveStruct({
	name: `${$moduleName}::State`,
	fields: {
		id: bcs.Address,
		governance_data_source: DataSource,
		stale_price_threshold: bcs.u64(),
		base_update_fee: bcs.u64(),
		fee_recipient_address: bcs.Address,
		last_executed_governance_sequence: bcs.u64(),
		consumed_vaas: ConsumedVAAs,
		upgrade_cap: UpgradeCap
	}
});
const CurrentDigest = new MoveStruct({
	name: `${$moduleName}::CurrentDigest`,
	fields: { dummy_field: bcs.bool() }
});

//#endregion
export { State };
