// @mysten/deepbook-v3/dist/transactions/balanceManager.mjs
import { convertQuantity } from "../utils/conversion.mjs";
import { coinWithBalance } from "@mysten/sui/transactions";

//#region src/transactions/balanceManager.ts
/**
* BalanceManagerContract class for managing BalanceManager operations.
*/
var BalanceManagerContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for BalanceManagerContract
	*/
	constructor(config) {
		this.createAndShareBalanceManager = () => (tx) => {
			const manager = tx.moveCall({ target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::new` });
			tx.moveCall({
				target: "0x2::transfer::public_share_object",
				arguments: [manager],
				typeArguments: [`${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::BalanceManager`]
			});
		};
		this.createBalanceManagerWithOwner = (ownerAddress) => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::new_with_custom_owner`,
				arguments: [tx.pure.address(ownerAddress)]
			});
		};
		this.shareBalanceManager = (manager) => (tx) => {
			tx.moveCall({
				target: "0x2::transfer::public_share_object",
				arguments: [manager],
				typeArguments: [`${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::BalanceManager`]
			});
		};
		this.depositIntoManager = (managerKey, coinKey, amountToDeposit) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const managerId = this.#config.getBalanceManager(managerKey).address;
			const coin = this.#config.getCoin(coinKey);
			const depositInput = convertQuantity(amountToDeposit, coin.scalar);
			const deposit = coinWithBalance({
				type: coin.type,
				balance: depositInput
			});
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::deposit`,
				arguments: [tx.object(managerId), deposit],
				typeArguments: [coin.type]
			});
		};
		this.withdrawFromManager = (managerKey, coinKey, amountToWithdraw, recipient) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			const coin = this.#config.getCoin(coinKey);
			const withdrawInput = convertQuantity(amountToWithdraw, coin.scalar);
			const coinObject = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::withdraw`,
				arguments: [tx.object(managerId), tx.pure.u64(withdrawInput)],
				typeArguments: [coin.type]
			});
			tx.transferObjects([coinObject], recipient);
		};
		this.withdrawAllFromManager = (managerKey, coinKey, recipient) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			const coin = this.#config.getCoin(coinKey);
			const withdrawalCoin = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::withdraw_all`,
				arguments: [tx.object(managerId)],
				typeArguments: [coin.type]
			});
			tx.transferObjects([withdrawalCoin], recipient);
		};
		this.checkManagerBalance = (managerKey, coinKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			const coin = this.#config.getCoin(coinKey);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::balance`,
				arguments: [tx.object(managerId)],
				typeArguments: [coin.type]
			});
		};
		this.generateProof = (managerKey) => (tx) => {
			const balanceManager = this.#config.getBalanceManager(managerKey);
			return tx.add(balanceManager.tradeCap ? this.generateProofAsTrader(balanceManager.address, balanceManager.tradeCap) : this.generateProofAsOwner(balanceManager.address));
		};
		this.generateProofAsOwner = (managerId) => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::generate_proof_as_owner`,
				arguments: [tx.object(managerId)]
			});
		};
		this.generateProofAsTrader = (managerId, tradeCapId) => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::generate_proof_as_trader`,
				arguments: [tx.object(managerId), tx.object(tradeCapId)]
			});
		};
		this.mintTradeCap = (managerKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::mint_trade_cap`,
				arguments: [tx.object(managerId)]
			});
		};
		this.mintDepositCap = (managerKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::mint_deposit_cap`,
				arguments: [tx.object(managerId)]
			});
		};
		this.mintWithdrawalCap = (managerKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::mint_withdraw_cap`,
				arguments: [tx.object(managerId)]
			});
		};
		this.depositWithCap = (managerKey, coinKey, amountToDeposit) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const manager = this.#config.getBalanceManager(managerKey);
			const managerId = manager.address;
			if (!manager.depositCap) throw new Error(`DepositCap not set for ${managerKey}`);
			const depositCapId = manager.depositCap;
			const coin = this.#config.getCoin(coinKey);
			const depositInput = convertQuantity(amountToDeposit, coin.scalar);
			const deposit = coinWithBalance({
				type: coin.type,
				balance: depositInput
			});
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::deposit_with_cap`,
				arguments: [
					tx.object(managerId),
					tx.object(depositCapId),
					deposit
				],
				typeArguments: [coin.type]
			});
		};
		this.withdrawWithCap = (managerKey, coinKey, amountToWithdraw) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const manager = this.#config.getBalanceManager(managerKey);
			const managerId = manager.address;
			if (!manager.withdrawCap) throw new Error(`WithdrawCap not set for ${managerKey}`);
			const withdrawCapId = manager.withdrawCap;
			const coin = this.#config.getCoin(coinKey);
			const withdrawAmount = convertQuantity(amountToWithdraw, coin.scalar);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::withdraw_with_cap`,
				arguments: [
					tx.object(managerId),
					tx.object(withdrawCapId),
					tx.pure.u64(withdrawAmount)
				],
				typeArguments: [coin.type]
			});
		};
		this.setBalanceManagerReferral = (managerKey, referral, tradeCap) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::set_balance_manager_referral`,
				arguments: [
					tx.object(managerId),
					tx.object(referral),
					tradeCap
				]
			});
		};
		this.unsetBalanceManagerReferral = (managerKey, poolKey, tradeCap) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			const poolId = this.#config.getPool(poolKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::unset_balance_manager_referral`,
				arguments: [
					tx.object(managerId),
					tx.pure.id(poolId),
					tradeCap
				]
			});
		};
		this.registerBalanceManager = (managerKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::register_balance_manager`,
				arguments: [tx.object(managerId), tx.object(this.#config.REGISTRY_ID)]
			});
		};
		this.owner = (managerKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::owner`,
				arguments: [tx.object(managerId)]
			});
		};
		this.id = (managerKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::id`,
				arguments: [tx.object(managerId)]
			});
		};
		this.balanceManagerReferralOwner = (referralId) => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::balance_manager_referral_owner`,
				arguments: [tx.object(referralId)]
			});
		};
		this.balanceManagerReferralPoolId = (referralId) => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::balance_manager_referral_pool_id`,
				arguments: [tx.object(referralId)]
			});
		};
		this.getBalanceManagerReferralId = (managerKey, poolKey) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			const poolId = this.#config.getPool(poolKey).address;
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::get_balance_manager_referral_id`,
				arguments: [tx.object(managerId), tx.pure.id(poolId)]
			});
		};
		this.revokeTradeCap = (managerKey, tradeCapId) => (tx) => {
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::balance_manager::revoke_trade_cap`,
				arguments: [tx.object(managerId), tx.pure.id(tradeCapId)]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { BalanceManagerContract };
