// @mysten/deepbook-v3/dist/transactions/deepbook.mjs
import { convertPrice, convertQuantity, convertRate } from "../utils/conversion.mjs";
import { DEEP_SCALAR, FLOAT_SCALAR, GAS_BUDGET, MAX_TIMESTAMP, POOL_CREATION_FEE_DEEP } from "../utils/config.mjs";
import { OrderType, SelfMatchingOptions } from "../types/index.mjs";
import { coinWithBalance } from "@mysten/sui/transactions";

//#region src/transactions/deepbook.ts
/**
* DeepBookContract class for managing DeepBook operations.
*/
var DeepBookContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for DeepBookContract
	*/
	constructor(config) {
		this.placeLimitOrder = (params) => (tx) => {
			const { poolKey, balanceManagerKey, clientOrderId, price, quantity, isBid, expiration = MAX_TIMESTAMP, orderType = OrderType.NO_RESTRICTION, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, payWithDeep = true } = params;
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::place_limit_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u64(clientOrderId),
					tx.pure.u8(orderType),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.pure.u64(expiration),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.placeMarketOrder = (params) => (tx) => {
			const { poolKey, balanceManagerKey, clientOrderId, quantity, isBid, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, payWithDeep = true } = params;
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::place_market_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u64(clientOrderId),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.modifyOrder = (poolKey, balanceManagerKey, orderId, newQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			const inputQuantity = convertQuantity(newQuantity, baseCoin.scalar);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::modify_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u128(orderId),
					tx.pure.u64(inputQuantity),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelOrder = (poolKey, balanceManagerKey, orderId) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::cancel_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u128(orderId),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelOrders = (poolKey, balanceManagerKey, orderIds) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::cancel_orders`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.vector("u128", orderIds),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelLiveOrder = (poolKey, balanceManagerKey, orderId) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::cancel_live_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u128(orderId),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelLiveOrders = (poolKey, balanceManagerKey, orderIds) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::cancel_live_orders`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.vector("u128", orderIds),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelAllOrders = (poolKey, balanceManagerKey) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::cancel_all_orders`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.withdrawSettledAmounts = (poolKey, balanceManagerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::withdraw_settled_amounts`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.withdrawSettledAmountsPermissionless = (poolKey, balanceManagerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::withdraw_settled_amounts_permissionless`,
				arguments: [tx.object(pool.address), tx.object(balanceManager.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.withdrawSettledAmountsManagerID = (poolKey, balanceManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::withdraw_settled_amounts_permissionless`,
				arguments: [tx.object(pool.address), tx.object(balanceManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.addDeepPricePoint = (targetPoolKey, referencePoolKey) => (tx) => {
			const targetPool = this.#config.getPool(targetPoolKey);
			const referencePool = this.#config.getPool(referencePoolKey);
			const targetBaseCoin = this.#config.getCoin(targetPool.baseCoin);
			const targetQuoteCoin = this.#config.getCoin(targetPool.quoteCoin);
			const referenceBaseCoin = this.#config.getCoin(referencePool.baseCoin);
			const referenceQuoteCoin = this.#config.getCoin(referencePool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::add_deep_price_point`,
				arguments: [
					tx.object(targetPool.address),
					tx.object(referencePool.address),
					tx.object.clock()
				],
				typeArguments: [
					targetBaseCoin.type,
					targetQuoteCoin.type,
					referenceBaseCoin.type,
					referenceQuoteCoin.type
				]
			});
		};
		this.claimRebates = (poolKey, balanceManagerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::claim_rebates`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.mintReferral = (poolKey, multiplier) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const adjustedNumber = convertRate(multiplier, FLOAT_SCALAR);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::mint_referral`,
				arguments: [tx.object(pool.address), tx.pure.u64(adjustedNumber)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.updatePoolReferralMultiplier = (poolKey, referral, multiplier) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const adjustedNumber = convertRate(multiplier, FLOAT_SCALAR);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::update_pool_referral_multiplier`,
				arguments: [
					tx.object(pool.address),
					tx.object(referral),
					tx.pure.u64(adjustedNumber)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.claimPoolReferralRewards = (poolKey, referral) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const [baseRewards, quoteRewards, deepRewards] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::claim_pool_referral_rewards`,
				arguments: [tx.object(pool.address), tx.object(referral)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return {
				baseRewards,
				quoteRewards,
				deepRewards
			};
		};
		this.updatePoolAllowedVersions = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::update_pool_allowed_versions`,
				arguments: [tx.object(pool.address), tx.object(this.#config.REGISTRY_ID)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getOrder = (poolKey, orderId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_order`,
				arguments: [tx.object(pool.address), tx.pure.u128(orderId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getOrders = (poolKey, orderIds) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_orders`,
				arguments: [tx.object(pool.address), tx.pure.vector("u128", orderIds)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.burnDeep = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::burn_deep`,
				arguments: [tx.object(pool.address), tx.object(this.#config.DEEP_TREASURY_ID)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.midPrice = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::mid_price`,
				arguments: [tx.object(pool.address), tx.object.clock()],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.whitelisted = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::whitelisted`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getQuoteQuantityOut = (poolKey, baseQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_quote_quantity_out`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(baseQuantity, baseCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getBaseQuantityOut = (poolKey, quoteQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const quoteScalar = quoteCoin.scalar;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_base_quantity_out`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(quoteQuantity, quoteScalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getQuantityOut = (poolKey, baseQuantity, quoteQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const quoteScalar = quoteCoin.scalar;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_quantity_out`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(baseQuantity, baseCoin.scalar)),
					tx.pure.u64(convertQuantity(quoteQuantity, quoteScalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.accountOpenOrders = (poolKey, managerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getBalanceManager(managerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::account_open_orders`,
				arguments: [tx.object(pool.address), tx.object(manager.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getLevel2Range = (poolKey, priceLow, priceHigh, isBid) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_level2_range`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertPrice(priceLow, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar)),
					tx.pure.u64(convertPrice(priceHigh, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar)),
					tx.pure.bool(isBid),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getLevel2TicksFromMid = (poolKey, tickFromMid) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_level2_ticks_from_mid`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(tickFromMid),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.vaultBalances = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::vault_balances`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getPoolIdByAssets = (baseType, quoteType) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_pool_id_by_asset`,
				arguments: [tx.object(this.#config.REGISTRY_ID)],
				typeArguments: [baseType, quoteType]
			});
		};
		this.swapExactBaseForQuote = (params) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			tx.setSenderIfNotSet(this.#config.address);
			if (params.quoteCoin) throw new Error("quoteCoin is not accepted for swapping base asset");
			const { poolKey, amount: baseAmount, deepAmount, minOut: minQuote } = params;
			const pool = this.#config.getPool(poolKey);
			const deepCoinType = this.#config.getCoin("DEEP").type;
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseCoinInput = params.baseCoin ?? coinWithBalance({
				type: baseCoin.type,
				balance: convertQuantity(baseAmount, baseCoin.scalar)
			});
			const deepCoin = params.deepCoin ?? coinWithBalance({
				type: deepCoinType,
				balance: convertQuantity(deepAmount, DEEP_SCALAR)
			});
			const minQuoteInput = convertQuantity(minQuote, quoteCoin.scalar);
			const [baseCoinResult, quoteCoinResult, deepCoinResult] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::swap_exact_base_for_quote`,
				arguments: [
					tx.object(pool.address),
					baseCoinInput,
					deepCoin,
					tx.pure.u64(minQuoteInput),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return [
				baseCoinResult,
				quoteCoinResult,
				deepCoinResult
			];
		};
		this.swapExactQuoteForBase = (params) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			tx.setSenderIfNotSet(this.#config.address);
			if (params.baseCoin) throw new Error("baseCoin is not accepted for swapping quote asset");
			const { poolKey, amount: quoteAmount, deepAmount, minOut: minBase } = params;
			const pool = this.#config.getPool(poolKey);
			const deepCoinType = this.#config.getCoin("DEEP").type;
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const quoteCoinInput = params.quoteCoin ?? coinWithBalance({
				type: quoteCoin.type,
				balance: convertQuantity(quoteAmount, quoteCoin.scalar)
			});
			const deepCoin = params.deepCoin ?? coinWithBalance({
				type: deepCoinType,
				balance: convertQuantity(deepAmount, DEEP_SCALAR)
			});
			const minBaseInput = convertQuantity(minBase, baseCoin.scalar);
			const [baseCoinResult, quoteCoinResult, deepCoinResult] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::swap_exact_quote_for_base`,
				arguments: [
					tx.object(pool.address),
					quoteCoinInput,
					deepCoin,
					tx.pure.u64(minBaseInput),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return [
				baseCoinResult,
				quoteCoinResult,
				deepCoinResult
			];
		};
		this.swapExactQuantity = (params) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			tx.setSenderIfNotSet(this.#config.address);
			const { poolKey, amount, deepAmount, minOut, baseCoin, quoteCoin, deepCoin, isBaseToCoin } = params;
			const pool = this.#config.getPool(poolKey);
			const deepCoinType = this.#config.getCoin("DEEP").type;
			const baseCoinType = this.#config.getCoin(pool.baseCoin);
			const quoteCoinType = this.#config.getCoin(pool.quoteCoin);
			const baseCoinInput = isBaseToCoin ? baseCoin ?? coinWithBalance({
				type: baseCoinType.type,
				balance: convertQuantity(amount, baseCoinType.scalar)
			}) : coinWithBalance({
				type: baseCoinType.type,
				balance: 0
			});
			const quoteCoinInput = isBaseToCoin ? coinWithBalance({
				type: quoteCoinType.type,
				balance: 0
			}) : quoteCoin ?? coinWithBalance({
				type: quoteCoinType.type,
				balance: convertQuantity(amount, quoteCoinType.scalar)
			});
			const deepCoinInput = deepCoin ?? coinWithBalance({
				type: deepCoinType,
				balance: convertQuantity(deepAmount, DEEP_SCALAR)
			});
			const minOutInput = convertQuantity(minOut, isBaseToCoin ? quoteCoinType.scalar : baseCoinType.scalar);
			const [baseCoinResult, quoteCoinResult, deepCoinResult] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::swap_exact_quantity`,
				arguments: [
					tx.object(pool.address),
					baseCoinInput,
					quoteCoinInput,
					deepCoinInput,
					tx.pure.u64(minOutInput),
					tx.object.clock()
				],
				typeArguments: [baseCoinType.type, quoteCoinType.type]
			});
			return [
				baseCoinResult,
				quoteCoinResult,
				deepCoinResult
			];
		};
		this.swapExactBaseForQuoteWithManager = (params) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const { poolKey, balanceManagerKey, tradeCap, depositCap, withdrawCap, amount: baseAmount, minOut: minQuote, baseCoin } = params;
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoinType = this.#config.getCoin(pool.baseCoin);
			const quoteCoinType = this.#config.getCoin(pool.quoteCoin);
			const baseCoinInput = baseCoin ?? coinWithBalance({
				type: baseCoinType.type,
				balance: convertQuantity(baseAmount, baseCoinType.scalar)
			});
			const minQuoteInput = convertQuantity(minQuote, quoteCoinType.scalar);
			const [baseCoinResult, quoteCoinResult] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::swap_exact_base_for_quote_with_manager`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tx.object(tradeCap),
					tx.object(depositCap),
					tx.object(withdrawCap),
					baseCoinInput,
					tx.pure.u64(minQuoteInput),
					tx.object.clock()
				],
				typeArguments: [baseCoinType.type, quoteCoinType.type]
			});
			return [baseCoinResult, quoteCoinResult];
		};
		this.swapExactQuoteForBaseWithManager = (params) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const { poolKey, balanceManagerKey, tradeCap, depositCap, withdrawCap, amount: quoteAmount, minOut: minBase, quoteCoin } = params;
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoinType = this.#config.getCoin(pool.baseCoin);
			const quoteCoinType = this.#config.getCoin(pool.quoteCoin);
			const quoteCoinInput = quoteCoin ?? coinWithBalance({
				type: quoteCoinType.type,
				balance: convertQuantity(quoteAmount, quoteCoinType.scalar)
			});
			const minBaseInput = convertQuantity(minBase, baseCoinType.scalar);
			const [baseCoinResult, quoteCoinResult] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::swap_exact_quote_for_base_with_manager`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tx.object(tradeCap),
					tx.object(depositCap),
					tx.object(withdrawCap),
					quoteCoinInput,
					tx.pure.u64(minBaseInput),
					tx.object.clock()
				],
				typeArguments: [baseCoinType.type, quoteCoinType.type]
			});
			return [baseCoinResult, quoteCoinResult];
		};
		this.swapExactQuantityWithManager = (params) => (tx) => {
			tx.setGasBudgetIfNotSet(GAS_BUDGET);
			const { poolKey, balanceManagerKey, tradeCap, depositCap, withdrawCap, amount, minOut, baseCoin, quoteCoin, isBaseToCoin } = params;
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoinType = this.#config.getCoin(pool.baseCoin);
			const quoteCoinType = this.#config.getCoin(pool.quoteCoin);
			const baseCoinInput = isBaseToCoin ? baseCoin ?? coinWithBalance({
				type: baseCoinType.type,
				balance: convertQuantity(amount, baseCoinType.scalar)
			}) : coinWithBalance({
				type: baseCoinType.type,
				balance: 0
			});
			const quoteCoinInput = isBaseToCoin ? coinWithBalance({
				type: quoteCoinType.type,
				balance: 0
			}) : quoteCoin ?? coinWithBalance({
				type: quoteCoinType.type,
				balance: convertQuantity(amount, quoteCoinType.scalar)
			});
			const minOutInput = convertQuantity(minOut, isBaseToCoin ? quoteCoinType.scalar : baseCoinType.scalar);
			const [baseCoinResult, quoteCoinResult] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::swap_exact_quantity_with_manager`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tx.object(tradeCap),
					tx.object(depositCap),
					tx.object(withdrawCap),
					baseCoinInput,
					quoteCoinInput,
					tx.pure.u64(minOutInput),
					tx.object.clock()
				],
				typeArguments: [baseCoinType.type, quoteCoinType.type]
			});
			return [baseCoinResult, quoteCoinResult];
		};
		this.createPermissionlessPool = (params) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const { baseCoinKey, quoteCoinKey, tickSize, lotSize, minSize, deepCoin } = params;
			const baseCoin = this.#config.getCoin(baseCoinKey);
			const quoteCoin = this.#config.getCoin(quoteCoinKey);
			const deepCoinType = this.#config.getCoin("DEEP").type;
			const baseScalar = baseCoin.scalar;
			const quoteScalar = quoteCoin.scalar;
			const adjustedTickSize = convertPrice(tickSize, FLOAT_SCALAR, quoteScalar, baseScalar);
			const adjustedLotSize = convertQuantity(lotSize, baseScalar);
			const adjustedMinSize = convertQuantity(minSize, baseScalar);
			const deepCoinInput = deepCoin ?? coinWithBalance({
				type: deepCoinType,
				balance: POOL_CREATION_FEE_DEEP
			});
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::create_permissionless_pool`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.pure.u64(adjustedTickSize),
					tx.pure.u64(adjustedLotSize),
					tx.pure.u64(adjustedMinSize),
					deepCoinInput
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.poolTradeParams = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::pool_trade_params`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.poolBookParams = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::pool_book_params`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.account = (poolKey, managerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::account`,
				arguments: [tx.object(pool.address), tx.object(managerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.lockedBalance = (poolKey, managerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const managerId = this.#config.getBalanceManager(managerKey).address;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::locked_balance`,
				arguments: [tx.object(pool.address), tx.object(managerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getPoolDeepPrice = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_order_deep_price`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getBalanceManagerIds = (owner) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::get_balance_manager_ids`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.pure.address(owner)]
			});
		};
		this.getPoolReferralBalances = (poolKey, referral) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_pool_referral_balances`,
				arguments: [tx.object(pool.address), tx.object(referral)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.poolReferralMultiplier = (poolKey, referral) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::pool_referral_multiplier`,
				arguments: [tx.object(pool.address), tx.object(referral)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.stablePool = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::stable_pool`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.registeredPool = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::registered_pool`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getQuoteQuantityOutInputFee = (poolKey, baseQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_quote_quantity_out_input_fee`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(baseQuantity, baseCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getBaseQuantityOutInputFee = (poolKey, quoteQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_base_quantity_out_input_fee`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(quoteQuantity, quoteCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getQuantityOutInputFee = (poolKey, baseQuantity, quoteQuantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_quantity_out_input_fee`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(baseQuantity, baseCoin.scalar)),
					tx.pure.u64(convertQuantity(quoteQuantity, quoteCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getBaseQuantityIn = (poolKey, targetQuoteQuantity, payWithDeep) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_base_quantity_in`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(targetQuoteQuantity, quoteCoin.scalar)),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getQuoteQuantityIn = (poolKey, targetBaseQuantity, payWithDeep) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_quote_quantity_in`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(targetBaseQuantity, baseCoin.scalar)),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getAccountOrderDetails = (poolKey, managerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getBalanceManager(managerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_account_order_details`,
				arguments: [tx.object(pool.address), tx.object(manager.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getOrderDeepRequired = (poolKey, baseQuantity, price) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(baseQuantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::get_order_deep_required`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(inputQuantity),
					tx.pure.u64(inputPrice)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.accountExists = (poolKey, managerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getBalanceManager(managerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::account_exists`,
				arguments: [tx.object(pool.address), tx.object(manager.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.poolTradeParamsNext = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::pool_trade_params_next`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.quorum = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::quorum`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.poolId = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::id`,
				arguments: [tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.canPlaceLimitOrder = (params) => (tx) => {
			const { poolKey, balanceManagerKey, price, quantity, isBid, payWithDeep, expireTimestamp } = params;
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::can_place_limit_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(manager.address),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.pure.u64(expireTimestamp),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.canPlaceMarketOrder = (params) => (tx) => {
			const { poolKey, balanceManagerKey, quantity, isBid, payWithDeep } = params;
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getBalanceManager(balanceManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::can_place_market_order`,
				arguments: [
					tx.object(pool.address),
					tx.object(manager.address),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.checkMarketOrderParams = (poolKey, quantity) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::check_market_order_params`,
				arguments: [tx.object(pool.address), tx.pure.u64(inputQuantity)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.checkLimitOrderParams = (poolKey, price, quantity, expireTimestamp) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::check_limit_order_params`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.u64(expireTimestamp),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { DeepBookContract };
