// @mysten/deepbook-v3/dist/transactions/deepbookAdmin.mjs
import { convertPrice, convertQuantity, convertRate } from "../utils/conversion.mjs";
import { FLOAT_SCALAR } from "../utils/config.mjs";

//#region src/transactions/deepbookAdmin.ts
/**
* DeepBookAdminContract class for managing admin actions.
*/
var DeepBookAdminContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for DeepBookAdminContract
	*/
	constructor(config) {
		this.createPoolAdmin = (params) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const { baseCoinKey, quoteCoinKey, tickSize, lotSize, minSize, whitelisted, stablePool } = params;
			const baseCoin = this.#config.getCoin(baseCoinKey);
			const quoteCoin = this.#config.getCoin(quoteCoinKey);
			const baseScalar = baseCoin.scalar;
			const quoteScalar = quoteCoin.scalar;
			const adjustedTickSize = convertPrice(tickSize, FLOAT_SCALAR, quoteScalar, baseScalar);
			const adjustedLotSize = convertQuantity(lotSize, baseScalar);
			const adjustedMinSize = convertQuantity(minSize, baseScalar);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::create_pool_admin`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.pure.u64(adjustedTickSize),
					tx.pure.u64(adjustedLotSize),
					tx.pure.u64(adjustedMinSize),
					tx.pure.bool(whitelisted),
					tx.pure.bool(stablePool),
					tx.object(this.#adminCap())
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.unregisterPoolAdmin = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::unregister_pool_admin`,
				arguments: [
					tx.object(pool.address),
					tx.object(this.#config.REGISTRY_ID),
					tx.object(this.#adminCap())
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.updateAllowedVersions = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::update_allowed_versions`,
				arguments: [
					tx.object(pool.address),
					tx.object(this.#config.REGISTRY_ID),
					tx.object(this.#adminCap())
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.enableVersion = (version) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::enable_version`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.pure.u64(version),
					tx.object(this.#adminCap())
				]
			});
		};
		this.disableVersion = (version) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::disable_version`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.pure.u64(version),
					tx.object(this.#adminCap())
				]
			});
		};
		this.setTreasuryAddress = (treasuryAddress) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::set_treasury_address`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.pure.address(treasuryAddress),
					tx.object(this.#adminCap())
				]
			});
		};
		this.addStableCoin = (stableCoinKey) => (tx) => {
			const stableCoinType = this.#config.getCoin(stableCoinKey).type;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::add_stablecoin`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.object(this.#adminCap())],
				typeArguments: [stableCoinType]
			});
		};
		this.removeStableCoin = (stableCoinKey) => (tx) => {
			const stableCoinType = this.#config.getCoin(stableCoinKey).type;
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::remove_stablecoin`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.object(this.#adminCap())],
				typeArguments: [stableCoinType]
			});
		};
		this.adjustTickSize = (poolKey, newTickSize) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseScalar = baseCoin.scalar;
			const quoteScalar = quoteCoin.scalar;
			const adjustedTickSize = convertPrice(newTickSize, FLOAT_SCALAR, quoteScalar, baseScalar);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::adjust_tick_size_admin`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(adjustedTickSize),
					tx.object(this.#adminCap()),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.adjustMinLotSize = (poolKey, newLotSize, newMinSize) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseScalar = baseCoin.scalar;
			const adjustedLotSize = convertQuantity(newLotSize, baseScalar);
			const adjustedMinSize = convertQuantity(newMinSize, baseScalar);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::adjust_min_lot_size_admin`,
				arguments: [
					tx.object(pool.address),
					tx.pure.u64(adjustedLotSize),
					tx.pure.u64(adjustedMinSize),
					tx.object(this.#adminCap()),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.initBalanceManagerMap = () => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::init_balance_manager_map`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.object(this.#adminCap())]
			});
		};
		this.setEwmaParams = (poolKey, params) => (tx) => {
			const { alpha, zScoreThreshold, additionalTakerFee } = params;
			const adjustedAlpha = convertRate(alpha, FLOAT_SCALAR);
			const adjustedZScoreThreshold = convertRate(zScoreThreshold, FLOAT_SCALAR);
			const adjustedAdditionalTakerFee = convertRate(additionalTakerFee, FLOAT_SCALAR);
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::set_ewma_params`,
				arguments: [
					tx.object(pool.address),
					tx.object(this.#adminCap()),
					tx.pure.u64(adjustedAlpha),
					tx.pure.u64(adjustedZScoreThreshold),
					tx.pure.u64(adjustedAdditionalTakerFee),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.enableEwmaState = (poolKey, enable) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::enable_ewma_state`,
				arguments: [
					tx.object(pool.address),
					tx.object(this.#adminCap()),
					tx.pure.bool(enable),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.authorizeMarginApp = () => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::authorize_app`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.object(this.#adminCap())],
				typeArguments: [`${this.#config.MARGIN_V1}::margin_manager::MarginApp`]
			});
		};
		this.deauthorizeMarginApp = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::deauthorize_app`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.object(this.#adminCap())],
				typeArguments: [`${this.#config.MARGIN_V1}::margin_manager::MarginApp`]
			});
		};
		this.mintCorePauseCap = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::mint_pause_cap`,
				arguments: [tx.object(this.#config.REGISTRY_ID), tx.object(this.#adminCap())]
			});
		};
		this.revokeCorePauseCap = (pauseCapId) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::revoke_pause_cap`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.object(this.#adminCap()),
					tx.pure.id(pauseCapId)
				]
			});
		};
		this.disableVersionWithCorePauseCap = (version, pauseCapId) => (tx) => {
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::disable_version_pause_cap`,
				arguments: [
					tx.object(this.#config.REGISTRY_ID),
					tx.pure.u64(version),
					tx.object(pauseCapId)
				]
			});
		};
		this.corePauseCaps = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::allowed_pause_caps`,
				arguments: [tx.object(this.#config.REGISTRY_ID)]
			});
		};
		this.#config = config;
	}
	/**
	* @returns The admin capability required for admin operations
	* @throws Error if the admin capability is not set
	*/
	#adminCap() {
		const adminCap = this.#config.adminCap;
		if (!adminCap) throw new Error("ADMIN_CAP environment variable not set");
		return adminCap;
	}
};

//#endregion
export { DeepBookAdminContract };
