// @mysten/deepbook-v3/dist/transactions/flashLoans.mjs
import { convertQuantity } from "../utils/conversion.mjs";

//#region src/transactions/flashLoans.ts
/**
* FlashLoanContract class for managing flash loans.
*/
var FlashLoanContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration object for DeepBook
	*/
	constructor(config) {
		this.borrowBaseAsset = (poolKey, borrowAmount) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputQuantity = convertQuantity(borrowAmount, baseCoin.scalar);
			const [baseCoinResult, flashLoan] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::borrow_flashloan_base`,
				arguments: [tx.object(pool.address), tx.pure.u64(inputQuantity)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return [baseCoinResult, flashLoan];
		};
		this.returnBaseAsset = (poolKey, borrowAmount, baseCoinInput, flashLoan) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const borrowScalar = baseCoin.scalar;
			const [baseCoinReturn] = tx.splitCoins(baseCoinInput, [tx.pure.u64(convertQuantity(borrowAmount, borrowScalar))]);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::return_flashloan_base`,
				arguments: [
					tx.object(pool.address),
					baseCoinReturn,
					flashLoan
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return baseCoinInput;
		};
		this.borrowQuoteAsset = (poolKey, borrowAmount) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputQuantity = convertQuantity(borrowAmount, quoteCoin.scalar);
			const [quoteCoinResult, flashLoan] = tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::borrow_flashloan_quote`,
				arguments: [tx.object(pool.address), tx.pure.u64(inputQuantity)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return [quoteCoinResult, flashLoan];
		};
		this.returnQuoteAsset = (poolKey, borrowAmount, quoteCoinInput, flashLoan) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const borrowScalar = quoteCoin.scalar;
			const [quoteCoinReturn] = tx.splitCoins(quoteCoinInput, [tx.pure.u64(convertQuantity(borrowAmount, borrowScalar))]);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::return_flashloan_quote`,
				arguments: [
					tx.object(pool.address),
					quoteCoinReturn,
					flashLoan
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return quoteCoinInput;
		};
		this.#config = config;
	}
};

//#endregion
export { FlashLoanContract };
