// @mysten/deepbook-v3/dist/transactions/governance.mjs
import { convertQuantity, convertRate } from "../utils/conversion.mjs";
import { DEEP_SCALAR, FLOAT_SCALAR } from "../utils/config.mjs";

//#region src/transactions/governance.ts
/**
* GovernanceContract class for managing governance operations in DeepBook.
*/
var GovernanceContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for GovernanceContract
	*/
	constructor(config) {
		this.stake = (poolKey, balanceManagerKey, stakeAmount) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const stakeInput = convertQuantity(stakeAmount, DEEP_SCALAR);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::stake`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u64(stakeInput)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.unstake = (poolKey, balanceManagerKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::unstake`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.submitProposal = (params) => (tx) => {
			const { poolKey, balanceManagerKey, takerFee, makerFee, stakeRequired } = params;
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::submit_proposal`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.u64(convertRate(takerFee, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(makerFee, FLOAT_SCALAR)),
					tx.pure.u64(convertQuantity(stakeRequired, DEEP_SCALAR))
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.vote = (poolKey, balanceManagerKey, proposal_id) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const balanceManager = this.#config.getBalanceManager(balanceManagerKey);
			const tradeProof = tx.add(this.#config.balanceManager.generateProof(balanceManagerKey));
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::vote`,
				arguments: [
					tx.object(pool.address),
					tx.object(balanceManager.address),
					tradeProof,
					tx.pure.id(proposal_id)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { GovernanceContract };
