// @mysten/deepbook-v3/dist/transactions/marginTPSL.mjs
import { convertPrice, convertQuantity } from "../utils/conversion.mjs";
import { FLOAT_SCALAR, MAX_TIMESTAMP } from "../utils/config.mjs";
import { OrderType, SelfMatchingOptions } from "../types/index.mjs";

//#region src/transactions/marginTPSL.ts
/**
* MarginTPSLContract class for managing Take Profit / Stop Loss operations.
*/
var MarginTPSLContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginTPSLContract
	*/
	constructor(config) {
		this.newCondition = (poolKey, triggerBelowPrice, triggerPrice) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const inputPrice = convertPrice(triggerPrice, FLOAT_SCALAR, this.#config.getCoin(pool.quoteCoin).scalar, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::tpsl::new_condition`,
				arguments: [tx.pure.bool(triggerBelowPrice), tx.pure.u64(inputPrice)]
			});
		};
		this.newPendingLimitOrder = (poolKey, params) => (tx) => {
			const { clientOrderId, orderType = OrderType.NO_RESTRICTION, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, price, quantity, isBid, payWithDeep = true, expireTimestamp = MAX_TIMESTAMP } = params;
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, this.#config.getCoin(pool.quoteCoin).scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::tpsl::new_pending_limit_order`,
				arguments: [
					tx.pure.u64(clientOrderId),
					tx.pure.u8(orderType),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.pure.u64(expireTimestamp)
				]
			});
		};
		this.newPendingMarketOrder = (poolKey, params) => (tx) => {
			const { clientOrderId, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, quantity, isBid, payWithDeep = true } = params;
			const pool = this.#config.getPool(poolKey);
			const inputQuantity = convertQuantity(quantity, this.#config.getCoin(pool.baseCoin).scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::tpsl::new_pending_market_order`,
				arguments: [
					tx.pure.u64(clientOrderId),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep)
				]
			});
		};
		this.addConditionalOrder = (params) => (tx) => {
			const { marginManagerKey, conditionalOrderId, triggerBelowPrice, triggerPrice, pendingOrder } = params;
			const manager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const condition = this.newCondition(manager.poolKey, triggerBelowPrice, triggerPrice)(tx);
			const pending = "price" in pendingOrder ? this.newPendingLimitOrder(manager.poolKey, pendingOrder)(tx) : this.newPendingMarketOrder(manager.poolKey, pendingOrder)(tx);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::add_conditional_order`,
				arguments: [
					tx.object(manager.address),
					tx.object(pool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.u64(conditionalOrderId),
					condition,
					pending,
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelAllConditionalOrders = (marginManagerKey) => (tx) => {
			const manager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::cancel_all_conditional_orders`,
				arguments: [tx.object(manager.address), tx.object.clock()],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelConditionalOrder = (marginManagerKey, conditionalOrderId) => (tx) => {
			const manager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::cancel_conditional_order`,
				arguments: [
					tx.object(manager.address),
					tx.pure.u64(conditionalOrderId),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.executeConditionalOrders = (managerAddress, poolKey, maxOrdersToExecute) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::execute_conditional_orders_v2`,
				arguments: [
					tx.object(managerAddress),
					tx.object(pool.address),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.u64(maxOrdersToExecute),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.conditionalOrderIds = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::conditional_order_ids`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.conditionalOrder = (poolKey, marginManagerId, conditionalOrderId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::conditional_order`,
				arguments: [tx.object(marginManagerId), tx.pure.u64(conditionalOrderId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.lowestTriggerAbovePrice = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::lowest_trigger_above_price`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.highestTriggerBelowPrice = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::highest_trigger_below_price`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { MarginTPSLContract };
