// @mysten/deepbook-v3/dist/transactions/marginMaintainer.mjs
import { convertQuantity, convertRate } from "../utils/conversion.mjs";
import { FLOAT_SCALAR } from "../utils/config.mjs";

//#region src/transactions/marginMaintainer.ts
/**
* DeepBookMaintainerContract class for managing maintainer actions.
*/
var MarginMaintainerContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginMaintainerContract
	*/
	constructor(config) {
		this.createMarginPool = (coinKey, poolConfig) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::create_margin_pool`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					poolConfig,
					tx.object(this.#marginMaintainerCap()),
					tx.object.clock()
				],
				typeArguments: [coin.type]
			});
		};
		this.newProtocolConfig = (coinKey, marginPoolConfig, interestConfig) => (tx) => {
			const marginPoolConfigObject = marginPoolConfig.rateLimitCapacity !== void 0 && marginPoolConfig.rateLimitRefillRatePerMs !== void 0 && marginPoolConfig.rateLimitEnabled !== void 0 ? this.newMarginPoolConfigWithRateLimit(coinKey, {
				...marginPoolConfig,
				rateLimitCapacity: marginPoolConfig.rateLimitCapacity,
				rateLimitRefillRatePerMs: marginPoolConfig.rateLimitRefillRatePerMs,
				rateLimitEnabled: marginPoolConfig.rateLimitEnabled
			})(tx) : this.newMarginPoolConfig(coinKey, marginPoolConfig)(tx);
			const interestConfigObject = this.newInterestConfig(interestConfig)(tx);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::protocol_config::new_protocol_config`,
				arguments: [marginPoolConfigObject, interestConfigObject]
			});
		};
		this.newMarginPoolConfig = (coinKey, marginPoolConfig) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			const { supplyCap, maxUtilizationRate, protocolSpread, minBorrow } = marginPoolConfig;
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::protocol_config::new_margin_pool_config`,
				arguments: [
					tx.pure.u64(convertQuantity(supplyCap, coin.scalar)),
					tx.pure.u64(convertRate(maxUtilizationRate, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(protocolSpread, FLOAT_SCALAR)),
					tx.pure.u64(convertQuantity(minBorrow, coin.scalar))
				]
			});
		};
		this.newMarginPoolConfigWithRateLimit = (coinKey, marginPoolConfig) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			const { supplyCap, maxUtilizationRate, protocolSpread, minBorrow, rateLimitCapacity, rateLimitRefillRatePerMs, rateLimitEnabled } = marginPoolConfig;
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::protocol_config::new_margin_pool_config_with_rate_limit`,
				arguments: [
					tx.pure.u64(convertQuantity(supplyCap, coin.scalar)),
					tx.pure.u64(convertRate(maxUtilizationRate, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(protocolSpread, FLOAT_SCALAR)),
					tx.pure.u64(convertQuantity(minBorrow, coin.scalar)),
					tx.pure.u64(convertQuantity(rateLimitCapacity, coin.scalar)),
					tx.pure.u64(convertQuantity(rateLimitRefillRatePerMs, coin.scalar)),
					tx.pure.bool(rateLimitEnabled)
				]
			});
		};
		this.newInterestConfig = (interestConfig) => (tx) => {
			const { baseRate, baseSlope, optimalUtilization, excessSlope } = interestConfig;
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::protocol_config::new_interest_config`,
				arguments: [
					tx.pure.u64(convertRate(baseRate, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(baseSlope, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(optimalUtilization, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(excessSlope, FLOAT_SCALAR))
				]
			});
		};
		this.enableDeepbookPoolForLoan = (deepbookPoolKey, coinKey, marginPoolCap) => (tx) => {
			const deepbookPool = this.#config.getPool(deepbookPoolKey);
			const marginPool = this.#config.getMarginPool(coinKey);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::enable_deepbook_pool_for_loan`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.id(deepbookPool.address),
					tx.object(marginPoolCap),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.disableDeepbookPoolForLoan = (deepbookPoolKey, coinKey, marginPoolCap) => (tx) => {
			const deepbookPool = this.#config.getPool(deepbookPoolKey);
			const marginPool = this.#config.getMarginPool(coinKey);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::disable_deepbook_pool_for_loan`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.id(deepbookPool.address),
					tx.object(marginPoolCap),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.updateInterestParams = (coinKey, marginPoolCap, interestConfig) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			const interestConfigObject = this.newInterestConfig(interestConfig)(tx);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::update_interest_params`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					interestConfigObject,
					tx.object(marginPoolCap),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.updateMarginPoolConfig = (coinKey, marginPoolCap, marginPoolConfig) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			const marginPoolConfigObject = marginPoolConfig.rateLimitCapacity !== void 0 && marginPoolConfig.rateLimitRefillRatePerMs !== void 0 && marginPoolConfig.rateLimitEnabled !== void 0 ? this.newMarginPoolConfigWithRateLimit(coinKey, {
				...marginPoolConfig,
				rateLimitCapacity: marginPoolConfig.rateLimitCapacity,
				rateLimitRefillRatePerMs: marginPoolConfig.rateLimitRefillRatePerMs,
				rateLimitEnabled: marginPoolConfig.rateLimitEnabled
			})(tx) : this.newMarginPoolConfig(coinKey, marginPoolConfig)(tx);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::update_margin_pool_config`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					marginPoolConfigObject,
					tx.object(marginPoolCap),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.#config = config;
	}
	/**
	* @returns The admin capability required for admin operations
	* @throws Error if the admin capability is not set
	*/
	#marginMaintainerCap() {
		const marginMaintainerCap = this.#config.marginMaintainerCap;
		if (!marginMaintainerCap) throw new Error("MARGIN_ADMIN_CAP environment variable not set");
		return marginMaintainerCap;
	}
};

//#endregion
export { MarginMaintainerContract };
