// @mysten/deepbook-v3/dist/transactions/marginRegistry.mjs
//#region src/transactions/marginRegistry.ts
/**
* MarginRegistryContract class for managing MarginRegistry read-only operations.
*/
var MarginRegistryContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginRegistryContract
	*/
	constructor(config) {
		this.poolEnabled = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::pool_enabled`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getMarginPoolId = (coinKey) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::get_margin_pool_id`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID)],
				typeArguments: [coin.type]
			});
		};
		this.getDeepbookPoolMarginPoolIds = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::get_deepbook_pool_margin_pool_ids`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.getMarginManagerIds = (owner) => (tx) => {
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::get_margin_manager_ids`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.address(owner)],
				typeArguments: []
			});
		};
		this.baseMarginPoolId = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::base_margin_pool_id`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.quoteMarginPoolId = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::quote_margin_pool_id`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.minWithdrawRiskRatio = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::min_withdraw_risk_ratio`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.minBorrowRiskRatio = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::min_borrow_risk_ratio`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.liquidationRiskRatio = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::liquidation_risk_ratio`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.targetLiquidationRiskRatio = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::target_liquidation_risk_ratio`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.userLiquidationReward = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::user_liquidation_reward`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.poolLiquidationReward = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::pool_liquidation_reward`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.id(pool.address)],
				typeArguments: []
			});
		};
		this.allowedMaintainers = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::allowed_maintainers`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID)],
				typeArguments: []
			});
		};
		this.allowedPauseCaps = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::allowed_pause_caps`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID)],
				typeArguments: []
			});
		};
		this.#config = config;
	}
};

//#endregion
export { MarginRegistryContract };
