// @mysten/deepbook-v3/dist/transactions/marginLiquidations.mjs
import { convertQuantity } from "../utils/conversion.mjs";
import { coinWithBalance } from "@mysten/sui/transactions";

//#region src/transactions/marginLiquidations.ts
/**
* MarginLiquidationsContract class for managing LiquidationVault operations.
*/
var MarginLiquidationsContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginLiquidationsContract
	*/
	constructor(config) {
		this.createLiquidationVault = (liquidationAdminCap) => (tx) => {
			tx.moveCall({
				target: `${this.#config.LIQUIDATION_PACKAGE_ID}::liquidation_vault::create_liquidation_vault`,
				arguments: [tx.object(liquidationAdminCap)]
			});
		};
		this.deposit = (vaultId, liquidationAdminCap, coinKey, amount) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			const depositCoin = coinWithBalance({
				type: coin.type,
				balance: convertQuantity(amount, coin.scalar)
			});
			tx.moveCall({
				target: `${this.#config.LIQUIDATION_PACKAGE_ID}::liquidation_vault::deposit`,
				arguments: [
					tx.object(vaultId),
					tx.object(liquidationAdminCap),
					depositCoin
				],
				typeArguments: [coin.type]
			});
		};
		this.withdraw = (vaultId, liquidationAdminCap, coinKey, amount) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			return tx.moveCall({
				target: `${this.#config.LIQUIDATION_PACKAGE_ID}::liquidation_vault::withdraw`,
				arguments: [
					tx.object(vaultId),
					tx.object(liquidationAdminCap),
					tx.pure.u64(convertQuantity(amount, coin.scalar))
				],
				typeArguments: [coin.type]
			});
		};
		this.liquidateBase = (vaultId, managerAddress, poolKey, repayAmount) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const repayAmountArg = repayAmount !== void 0 ? tx.pure.option("u64", convertQuantity(repayAmount, baseCoin.scalar)) : tx.pure.option("u64", null);
			tx.moveCall({
				target: `${this.#config.LIQUIDATION_PACKAGE_ID}::liquidation_vault::liquidate_base`,
				arguments: [
					tx.object(vaultId),
					tx.object(managerAddress),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(pool.address),
					repayAmountArg,
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.liquidateQuote = (vaultId, managerAddress, poolKey, repayAmount) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const repayAmountArg = repayAmount !== void 0 ? tx.pure.option("u64", convertQuantity(repayAmount, quoteCoin.scalar)) : tx.pure.option("u64", null);
			tx.moveCall({
				target: `${this.#config.LIQUIDATION_PACKAGE_ID}::liquidation_vault::liquidate_quote`,
				arguments: [
					tx.object(vaultId),
					tx.object(managerAddress),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(pool.address),
					repayAmountArg,
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.balance = (vaultId, coinKey) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			return tx.moveCall({
				target: `${this.#config.LIQUIDATION_PACKAGE_ID}::liquidation_vault::balance`,
				arguments: [tx.object(vaultId)],
				typeArguments: [coin.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { MarginLiquidationsContract };
