// @mysten/deepbook-v3/dist/transactions/marginAdmin.mjs
import { convertRate } from "../utils/conversion.mjs";
import { FLOAT_SCALAR } from "../utils/config.mjs";
import { hexToBytes } from "@noble/hashes/utils.js";

//#region src/transactions/marginAdmin.ts
/**
* MarginAdminContract class for managing admin actions.
*/
var MarginAdminContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginAdminContract
	*/
	constructor(config) {
		this.mintMaintainerCap = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::mint_maintainer_cap`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object.clock()
				]
			});
		};
		this.revokeMaintainerCap = (maintainerCapId) => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::revoke_maintainer_cap`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(maintainerCapId),
					tx.object.clock()
				]
			});
		};
		this.registerDeepbookPool = (poolKey, poolConfig) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::register_deepbook_pool`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					poolConfig,
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.enableDeepbookPool = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::enable_deepbook_pool`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.disableDeepbookPool = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::disable_deepbook_pool`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.updateRiskParams = (poolKey, poolConfig) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::update_risk_params`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					poolConfig,
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.setPriceTolerance = (poolKey, tolerance) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::set_price_tolerance`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					tx.pure.u64(convertRate(tolerance, FLOAT_SCALAR)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.setMaxPriceAge = (poolKey, maxAgeMs) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::set_max_price_age`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					tx.pure.u64(maxAgeMs),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.setMaxOrderTtl = (poolKey, maxOrderTtlMs) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::set_max_order_ttl`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object(pool.address),
					tx.pure.u64(maxOrderTtlMs),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.addConfig = (config$1) => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::add_config`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					config$1
				],
				typeArguments: [`${this.#config.MARGIN_PACKAGE_ID}::oracle::PythConfig`]
			});
		};
		this.removeConfig = () => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::remove_config`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.object(this.#marginAdminCap())],
				typeArguments: [`${this.#config.MARGIN_PACKAGE_ID}::oracle::PythConfig`]
			});
		};
		this.enableVersion = (version) => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::enable_version`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.u64(version),
					tx.object(this.#marginAdminCap())
				]
			});
		};
		this.disableVersion = (version) => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::disable_version`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.u64(version),
					tx.object(this.#marginAdminCap())
				]
			});
		};
		this.newPoolConfig = (poolKey, poolConfigParams) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const { minWithdrawRiskRatio, minBorrowRiskRatio, liquidationRiskRatio, targetLiquidationRiskRatio, userLiquidationReward, poolLiquidationReward } = poolConfigParams;
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::new_pool_config`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.u64(convertRate(minWithdrawRiskRatio, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(minBorrowRiskRatio, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(liquidationRiskRatio, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(targetLiquidationRiskRatio, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(userLiquidationReward, FLOAT_SCALAR)),
					tx.pure.u64(convertRate(poolLiquidationReward, FLOAT_SCALAR))
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.newPoolConfigWithLeverage = (poolKey, leverage) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::new_pool_config_with_leverage`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.pure.u64(convertRate(leverage, FLOAT_SCALAR))],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.newCoinTypeData = (coinKey, maxConfBps, maxEwmaDifferenceBps) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			if (!coin.feed) throw new Error("Coin feed not found");
			const priceFeedInput = new Uint8Array(hexToBytes(coin["feed"].startsWith("0x") ? coin.feed.slice(2) : coin["feed"]));
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::oracle::new_coin_type_data_from_currency`,
				arguments: [
					tx.object(coin.currencyId),
					tx.pure.vector("u8", priceFeedInput),
					tx.pure.u64(maxConfBps),
					tx.pure.u64(maxEwmaDifferenceBps)
				],
				typeArguments: [coin.type]
			});
		};
		this.newPythConfig = (coinSetups, maxAgeSeconds) => (tx) => {
			const coinTypeDataList = [];
			for (const setup of coinSetups) coinTypeDataList.push(this.newCoinTypeData(setup.coinKey, setup.maxConfBps, setup.maxEwmaDifferenceBps)(tx));
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::oracle::new_pyth_config`,
				arguments: [tx.makeMoveVec({
					elements: coinTypeDataList,
					type: `${this.#config.MARGIN_PACKAGE_ID}::oracle::CoinTypeData`
				}), tx.pure.u64(maxAgeSeconds)]
			});
		};
		this.mintPauseCap = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::mint_pause_cap`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object.clock()
				]
			});
		};
		this.revokePauseCap = (pauseCapId) => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::revoke_pause_cap`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap()),
					tx.object.clock(),
					tx.pure.id(pauseCapId)
				]
			});
		};
		this.disableVersionPauseCap = (version, pauseCapId) => (tx) => {
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_registry::disable_version_pause_cap`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.pure.u64(version),
					tx.object(pauseCapId)
				]
			});
		};
		this.adminWithdrawDefaultReferralFees = (coinKey) => (tx) => {
			const coin = this.#config.getCoin(coinKey);
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::admin_withdraw_default_referral_fees`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(this.#marginAdminCap())
				],
				typeArguments: [coin.type]
			});
		};
		this.#config = config;
	}
	/**
	* @returns The admin capability required for admin operations
	* @throws Error if the admin capability is not set
	*/
	#marginAdminCap() {
		const marginAdminCap = this.#config.marginAdminCap;
		if (!marginAdminCap) throw new Error("MARGIN_ADMIN_CAP environment variable not set");
		return marginAdminCap;
	}
};

//#endregion
export { MarginAdminContract };
