// @mysten/deepbook-v3/dist/transactions/marginManager.mjs
import { convertPrice, convertQuantity } from "../utils/conversion.mjs";
import { FLOAT_SCALAR } from "../utils/config.mjs";
import { coinWithBalance } from "@mysten/sui/transactions";

//#region src/transactions/marginManager.ts
/**
* MarginManagerContract class for managing MarginManager operations.
*/
var MarginManagerContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginManagerContract
	*/
	constructor(config) {
		this.newMarginManager = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::new`,
				arguments: [
					tx.object(pool.address),
					tx.object(this.#config.REGISTRY_ID),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.newMarginManagerWithInitializer = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const [manager, initializer] = tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::new_with_initializer`,
				arguments: [
					tx.object(pool.address),
					tx.object(this.#config.REGISTRY_ID),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
			return {
				manager,
				initializer
			};
		};
		this.shareMarginManager = (poolKey, manager, initializer) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::share`,
				arguments: [manager, initializer],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.registerMarginManager = (managerKey) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::register_margin_manager`,
				arguments: [tx.object(manager.address), tx.object(this.#config.MARGIN_REGISTRY_ID)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.unregisterMarginManager = (managerKey) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::unregister_margin_manager`,
				arguments: [tx.object(manager.address), tx.object(this.#config.MARGIN_REGISTRY_ID)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.depositDuringInitialization = (params) => (tx) => {
			const { manager, poolKey, coinType } = params;
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const depositCoin = this.#config.getCoin(coinType);
			const coin = "amount" in params && params.amount !== void 0 ? coinWithBalance({
				type: depositCoin.type,
				balance: convertQuantity(params.amount, depositCoin.scalar)
			}) : params.coin;
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::deposit`,
				arguments: [
					manager,
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					coin,
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					depositCoin.type
				]
			});
		};
		this.depositBase = (params) => (tx) => {
			const { managerKey } = params;
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const coin = "amount" in params && params.amount !== void 0 ? coinWithBalance({
				type: baseCoin.type,
				balance: convertQuantity(params.amount, baseCoin.scalar)
			}) : params.coin;
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::deposit`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					coin,
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					baseCoin.type
				]
			});
		};
		this.depositQuote = (params) => (tx) => {
			const { managerKey } = params;
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const coin = "amount" in params && params.amount !== void 0 ? coinWithBalance({
				type: quoteCoin.type,
				balance: convertQuantity(params.amount, quoteCoin.scalar)
			}) : params.coin;
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::deposit`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					coin,
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					quoteCoin.type
				]
			});
		};
		this.depositDeep = (params) => (tx) => {
			const { managerKey } = params;
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const deepCoin = this.#config.getCoin("DEEP");
			const coin = "amount" in params && params.amount !== void 0 ? coinWithBalance({
				type: deepCoin.type,
				balance: convertQuantity(params.amount, deepCoin.scalar)
			}) : params.coin;
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::deposit`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					coin,
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					deepCoin.type
				]
			});
		};
		this.withdrawBase = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::withdraw`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(amount, baseCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					baseCoin.type
				]
			});
		};
		this.withdrawQuote = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::withdraw`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(amount, quoteCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					quoteCoin.type
				]
			});
		};
		this.withdrawDeep = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const deepCoin = this.#config.getCoin("DEEP");
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::withdraw`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(amount, deepCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					deepCoin.type
				]
			});
		};
		this.borrowBase = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::borrow_base`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(amount, baseCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.borrowQuote = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::borrow_quote`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(pool.address),
					tx.pure.u64(convertQuantity(amount, quoteCoin.scalar)),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.repayBase = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::repay_base`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseMarginPool.address),
					tx.object.option({
						type: "u64",
						value: amount ? tx.pure.u64(convertQuantity(amount, baseCoin.scalar)) : null
					}),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.repayQuote = (managerKey, amount) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::repay_quote`,
				arguments: [
					tx.object(manager.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(quoteMarginPool.address),
					tx.object.option({
						type: "u64",
						value: amount ? tx.pure.u64(convertQuantity(amount, quoteCoin.scalar)) : null
					}),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.liquidate = (managerAddress, poolKey, debtIsBase, repayCoin) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const marginPool = debtIsBase ? baseMarginPool : quoteMarginPool;
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::liquidate`,
				arguments: [
					tx.object(managerAddress),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(marginPool.address),
					tx.object(pool.address),
					repayCoin,
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					debtIsBase ? baseCoin.type : quoteCoin.type
				]
			});
		};
		this.setMarginManagerReferral = (managerKey, referral) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(manager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::set_margin_manager_referral`,
				arguments: [tx.object(manager.address), tx.object(referral)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.unsetMarginManagerReferral = (managerKey, poolKey) => (tx) => {
			const manager = this.#config.getMarginManager(managerKey);
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::unset_margin_manager_referral`,
				arguments: [tx.object(manager.address), tx.pure.id(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.ownerByPoolKey = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::owner`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.deepbookPool = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::deepbook_pool`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.marginPoolId = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::margin_pool_id`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.borrowedShares = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::borrowed_shares`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.borrowedBaseShares = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::borrowed_base_shares`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.borrowedQuoteShares = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::borrowed_quote_shares`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.hasBaseDebt = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::has_base_debt`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.balanceManager = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::balance_manager`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.calculateAssets = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::calculate_assets`,
				arguments: [tx.object(marginManagerId), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.calculateDebts = (poolKey, coinKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const debtCoin = this.#config.getCoin(coinKey);
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::calculate_debts`,
				arguments: [
					tx.object(marginManagerId),
					tx.object(marginPool.address),
					tx.object.clock()
				],
				typeArguments: [
					baseCoin.type,
					quoteCoin.type,
					debtCoin.type
				]
			});
		};
		this.managerState = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::manager_state`,
				arguments: [
					tx.object(marginManagerId),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object(pool.address),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.baseBalance = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::base_balance`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.quoteBalance = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::quote_balance`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.deepBalance = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::deep_balance`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.balanceManagerId = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::balance_manager_id`,
				arguments: [tx.object(marginManagerId)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getBalanceManagerReferralId = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::get_balance_manager_referral_id`,
				arguments: [tx.object(marginManagerId), tx.pure.id(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.accountExists = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::account_exists`,
				arguments: [tx.object(marginManagerId), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.account = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::account`,
				arguments: [tx.object(marginManagerId), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.accountOpenOrders = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::account_open_orders`,
				arguments: [tx.object(marginManagerId), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.getAccountOrderDetails = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::get_account_order_details`,
				arguments: [tx.object(marginManagerId), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.lockedBalance = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::locked_balance`,
				arguments: [tx.object(marginManagerId), tx.object(pool.address)],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.canPlaceLimitOrder = (poolKey, marginManagerId, price, quantity, isBid, payWithDeep, expireTimestamp) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::can_place_limit_order`,
				arguments: [
					tx.object(marginManagerId),
					tx.object(pool.address),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.pure.u64(expireTimestamp),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.canPlaceMarketOrder = (poolKey, marginManagerId, quantity, isBid, payWithDeep) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_manager::can_place_market_order`,
				arguments: [
					tx.object(marginManagerId),
					tx.object(pool.address),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { MarginManagerContract };
