// @mysten/deepbook-v3/dist/transactions/marginPool.mjs
import { convertQuantity } from "../utils/conversion.mjs";
import { coinWithBalance } from "@mysten/sui/transactions";

//#region src/transactions/marginPool.ts
/**
* MarginPoolContract class for managing MarginPool operations.
*/
var MarginPoolContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for MarginPoolContract
	*/
	constructor(config) {
		this.mintSupplierCap = () => (tx) => {
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::mint_supplier_cap`,
				arguments: [tx.object(this.#config.MARGIN_REGISTRY_ID), tx.object.clock()]
			});
		};
		this.supplyToMarginPool = (coinKey, supplierCap, amountToDeposit, referralId) => (tx) => {
			tx.setSenderIfNotSet(this.#config.address);
			const marginPool = this.#config.getMarginPool(coinKey);
			const coin = this.#config.getCoin(coinKey);
			const depositInput = convertQuantity(amountToDeposit, coin.scalar);
			const supply = coinWithBalance({
				type: coin.type,
				balance: depositInput
			});
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::supply`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					supplierCap,
					supply,
					tx.object.option({
						type: "0x2::object::ID",
						value: referralId ? tx.pure.id(referralId) : null
					}),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.withdrawFromMarginPool = (coinKey, supplierCap, amountToWithdraw) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			const coin = this.#config.getCoin(coinKey);
			const withdrawInput = amountToWithdraw !== void 0 ? convertQuantity(amountToWithdraw, coin.scalar) : null;
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::withdraw`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					supplierCap,
					tx.pure.option("u64", withdrawInput),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.mintSupplyReferral = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::mint_supply_referral`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.withdrawReferralFees = (coinKey, referralId) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::withdraw_referral_fees`,
				arguments: [
					tx.object(marginPool.address),
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(referralId)
				],
				typeArguments: [marginPool.type]
			});
		};
		this.getId = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::id`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.deepbookPoolAllowed = (coinKey, deepbookPoolId) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::deepbook_pool_allowed`,
				arguments: [tx.object(marginPool.address), tx.pure.id(deepbookPoolId)],
				typeArguments: [marginPool.type]
			});
		};
		this.totalSupply = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::total_supply`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.supplyShares = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::supply_shares`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.totalBorrow = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::total_borrow`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.borrowShares = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::borrow_shares`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.lastUpdateTimestamp = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::last_update_timestamp`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.supplyCap = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::supply_cap`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.maxUtilizationRate = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::max_utilization_rate`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.protocolSpread = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::protocol_spread`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.minBorrow = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::min_borrow`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.interestRate = (coinKey) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::interest_rate`,
				arguments: [tx.object(marginPool.address)],
				typeArguments: [marginPool.type]
			});
		};
		this.userSupplyShares = (coinKey, supplierCapId) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::user_supply_shares`,
				arguments: [tx.object(marginPool.address), tx.pure.id(supplierCapId)],
				typeArguments: [marginPool.type]
			});
		};
		this.userSupplyAmount = (coinKey, supplierCapId) => (tx) => {
			const marginPool = this.#config.getMarginPool(coinKey);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::margin_pool::user_supply_amount`,
				arguments: [
					tx.object(marginPool.address),
					tx.pure.id(supplierCapId),
					tx.object.clock()
				],
				typeArguments: [marginPool.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { MarginPoolContract };
