// @mysten/deepbook-v3/dist/transactions/poolProxy.mjs
import { convertPrice, convertQuantity, convertRate } from "../utils/conversion.mjs";
import { FLOAT_SCALAR, MAX_TIMESTAMP } from "../utils/config.mjs";
import { OrderType, SelfMatchingOptions } from "../types/index.mjs";

//#region src/transactions/poolProxy.ts
/**
* PoolProxyContract class for managing PoolProxy operations.
*/
var PoolProxyContract = class {
	#config;
	/**
	* @param {DeepBookConfig} config Configuration for PoolProxyContract
	*/
	constructor(config) {
		this.placeLimitOrder = (params) => (tx) => {
			const { poolKey, marginManagerKey, clientOrderId, price, quantity, isBid, expiration = MAX_TIMESTAMP, orderType = OrderType.NO_RESTRICTION, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, payWithDeep = true } = params;
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getMarginManager(marginManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::place_limit_order_v2`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(manager.address),
					tx.object(pool.address),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.pure.u64(clientOrderId),
					tx.pure.u8(orderType),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.pure.u64(expiration),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.placeMarketOrder = (params) => (tx) => {
			const { poolKey, marginManagerKey, clientOrderId, quantity, isBid, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, payWithDeep = true } = params;
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getMarginManager(marginManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::place_market_order_v2`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(manager.address),
					tx.object(pool.address),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.pure.u64(clientOrderId),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.placeReduceOnlyLimitOrder = (params) => (tx) => {
			const { poolKey, marginManagerKey, clientOrderId, price, quantity, isBid, expiration = MAX_TIMESTAMP, orderType = OrderType.NO_RESTRICTION, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, payWithDeep = true } = params;
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getMarginManager(marginManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const inputPrice = convertPrice(price, FLOAT_SCALAR, quoteCoin.scalar, baseCoin.scalar);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::place_reduce_only_limit_order_v2`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(manager.address),
					tx.object(pool.address),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.pure.u64(clientOrderId),
					tx.pure.u8(orderType),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputPrice),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.pure.u64(expiration),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.placeReduceOnlyMarketOrder = (params) => (tx) => {
			const { poolKey, marginManagerKey, clientOrderId, quantity, isBid, selfMatchingOption = SelfMatchingOptions.SELF_MATCHING_ALLOWED, payWithDeep = true } = params;
			const pool = this.#config.getPool(poolKey);
			const manager = this.#config.getMarginManager(marginManagerKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const baseMarginPool = this.#config.getMarginPool(pool.baseCoin);
			const quoteMarginPool = this.#config.getMarginPool(pool.quoteCoin);
			const inputQuantity = convertQuantity(quantity, baseCoin.scalar);
			return tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::place_reduce_only_market_order_v2`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(manager.address),
					tx.object(pool.address),
					tx.object(baseMarginPool.address),
					tx.object(quoteMarginPool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.pure.u64(clientOrderId),
					tx.pure.u8(selfMatchingOption),
					tx.pure.u64(inputQuantity),
					tx.pure.bool(isBid),
					tx.pure.bool(payWithDeep),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.modifyOrder = (marginManagerKey, orderId, newQuantity) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const inputQuantity = convertQuantity(newQuantity, baseCoin.scalar);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::modify_order`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.pure.u128(orderId),
					tx.pure.u64(inputQuantity),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelOrder = (marginManagerKey, orderId) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::cancel_order`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.pure.u128(orderId),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelOrders = (marginManagerKey, orderIds) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::cancel_orders`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.pure.vector("u128", orderIds),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.cancelAllOrders = (marginManagerKey) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::cancel_all_orders`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.withdrawSettledAmounts = (marginManagerKey) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::withdraw_settled_amounts`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.stake = (marginManagerKey, stakeAmount) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const stakeInput = convertQuantity(stakeAmount, this.#config.getCoin("DEEP").scalar);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::stake`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.pure.u64(stakeInput)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.unstake = (marginManagerKey) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::unstake`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.submitProposal = (marginManagerKey, params) => (tx) => {
			const { takerFee, makerFee, stakeRequired } = params;
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			const stakeInput = convertRate(stakeRequired, FLOAT_SCALAR);
			const takerFeeInput = convertRate(takerFee, FLOAT_SCALAR);
			const makerFeeInput = convertRate(makerFee, FLOAT_SCALAR);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::submit_proposal`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.pure.u64(takerFeeInput),
					tx.pure.u64(makerFeeInput),
					tx.pure.u64(stakeInput)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.vote = (marginManagerKey, proposalId) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::vote`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address),
					tx.pure.id(proposalId)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.claimRebate = (marginManagerKey) => (tx) => {
			const marginManager = this.#config.getMarginManager(marginManagerKey);
			const pool = this.#config.getPool(marginManager.poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::claim_rebates`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManager.address),
					tx.object(pool.address)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.withdrawMarginSettledAmounts = (poolKey, marginManagerId) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::withdraw_settled_amounts_permissionless`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(marginManagerId),
					tx.object(pool.address)
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.updateCurrentPrice = (poolKey) => (tx) => {
			const pool = this.#config.getPool(poolKey);
			const baseCoin = this.#config.getCoin(pool.baseCoin);
			const quoteCoin = this.#config.getCoin(pool.quoteCoin);
			if (!baseCoin.priceInfoObjectId) throw new Error(`Missing priceInfoObjectId for ${pool.baseCoin}`);
			if (!quoteCoin.priceInfoObjectId) throw new Error(`Missing priceInfoObjectId for ${pool.quoteCoin}`);
			tx.moveCall({
				target: `${this.#config.MARGIN_PACKAGE_ID}::pool_proxy::update_current_price`,
				arguments: [
					tx.object(this.#config.MARGIN_REGISTRY_ID),
					tx.object(pool.address),
					tx.object(baseCoin.priceInfoObjectId),
					tx.object(quoteCoin.priceInfoObjectId),
					tx.object.clock()
				],
				typeArguments: [baseCoin.type, quoteCoin.type]
			});
		};
		this.#config = config;
	}
};

//#endregion
export { PoolProxyContract };
