// @mysten/deepbook-v3/dist/client.mjs
import { BalanceManagerContract } from "./transactions/balanceManager.mjs";
import { DeepBookConfig } from "./utils/config.mjs";
import { AccountQueries } from "./queries/accountQueries.mjs";
import { BalanceManagerQueries } from "./queries/balanceManagerQueries.mjs";
import { MarginManagerQueries } from "./queries/marginManagerQueries.mjs";
import { MarginPoolQueries } from "./queries/marginPoolQueries.mjs";
import { OrderQueries } from "./queries/orderQueries.mjs";
import { PoolQueries } from "./queries/poolQueries.mjs";
import { PriceFeedQueries } from "./queries/priceFeedQueries.mjs";
import { QuantityQueries } from "./queries/quantityQueries.mjs";
import { ReferralQueries } from "./queries/referralQueries.mjs";
import { RegistryQueries } from "./queries/registryQueries.mjs";
import { TPSLQueries } from "./queries/tpslQueries.mjs";
import { DeepBookContract } from "./transactions/deepbook.mjs";
import { DeepBookAdminContract } from "./transactions/deepbookAdmin.mjs";
import { FlashLoanContract } from "./transactions/flashLoans.mjs";
import { GovernanceContract } from "./transactions/governance.mjs";
import { MarginAdminContract } from "./transactions/marginAdmin.mjs";
import { MarginLiquidationsContract } from "./transactions/marginLiquidations.mjs";
import { MarginMaintainerContract } from "./transactions/marginMaintainer.mjs";
import { MarginManagerContract } from "./transactions/marginManager.mjs";
import { MarginPoolContract } from "./transactions/marginPool.mjs";
import { MarginRegistryContract } from "./transactions/marginRegistry.mjs";
import { MarginTPSLContract } from "./transactions/marginTPSL.mjs";
import { PoolProxyContract } from "./transactions/poolProxy.mjs";
import { normalizeSuiAddress } from "@mysten/sui/utils";

//#region src/client.ts
function deepbook({ name = "deepbook", ...options }) {
	return {
		name,
		register: (client) => {
			return new DeepBookClient({
				client,
				network: client.network,
				...options
			});
		}
	};
}
/**
* DeepBookClient class for managing DeepBook operations.
*/
var DeepBookClient = class {
	#balanceManagerQueries;
	#poolQueries;
	#quantityQueries;
	#orderQueries;
	#accountQueries;
	#referralQueries;
	#priceFeedQueries;
	#marginPoolQueries;
	#marginManagerQueries;
	#tpslQueries;
	#registryQueries;
	/**
	* Creates a new DeepBookClient instance
	*/
	constructor({ client, address, network, balanceManagers, marginManagers, coins, pools, adminCap, marginAdminCap, marginMaintainerCap, packageIds, pyth }) {
		const normalizedAddress = normalizeSuiAddress(address);
		const config = new DeepBookConfig({
			address: normalizedAddress,
			network,
			balanceManagers,
			marginManagers,
			coins,
			pools,
			adminCap,
			marginAdminCap,
			marginMaintainerCap,
			packageIds,
			pyth
		});
		this.balanceManager = new BalanceManagerContract(config);
		this.deepBook = new DeepBookContract(config);
		this.deepBookAdmin = new DeepBookAdminContract(config);
		this.flashLoans = new FlashLoanContract(config);
		this.governance = new GovernanceContract(config);
		this.marginAdmin = new MarginAdminContract(config);
		this.marginMaintainer = new MarginMaintainerContract(config);
		this.marginPool = new MarginPoolContract(config);
		this.marginManager = new MarginManagerContract(config);
		this.marginRegistry = new MarginRegistryContract(config);
		this.marginLiquidations = new MarginLiquidationsContract(config);
		this.poolProxy = new PoolProxyContract(config);
		this.marginTPSL = new MarginTPSLContract(config);
		const ctx = {
			client,
			config,
			address: normalizedAddress,
			balanceManager: this.balanceManager,
			deepBook: this.deepBook,
			marginManager: this.marginManager,
			marginPool: this.marginPool,
			marginRegistry: this.marginRegistry,
			marginTPSL: this.marginTPSL
		};
		this.#balanceManagerQueries = new BalanceManagerQueries(ctx);
		this.#poolQueries = new PoolQueries(ctx);
		this.#quantityQueries = new QuantityQueries(ctx);
		this.#orderQueries = new OrderQueries(ctx);
		this.#accountQueries = new AccountQueries(ctx);
		this.#referralQueries = new ReferralQueries(ctx);
		this.#priceFeedQueries = new PriceFeedQueries(ctx);
		this.#marginPoolQueries = new MarginPoolQueries(ctx);
		this.#marginManagerQueries = new MarginManagerQueries(ctx);
		this.#tpslQueries = new TPSLQueries(ctx);
		this.#registryQueries = new RegistryQueries(ctx);
	}
	checkManagerBalance(managerKey, coinKey) {
		return this.#balanceManagerQueries.checkManagerBalance(managerKey, coinKey);
	}
	checkManagerBalanceWithAddress(managerAddress, coinKey) {
		return this.#balanceManagerQueries.checkManagerBalanceWithAddress(managerAddress, coinKey);
	}
	checkManagerBalancesWithAddress(managerAddresses, coinKeys) {
		return this.#balanceManagerQueries.checkManagerBalancesWithAddress(managerAddresses, coinKeys);
	}
	getBalanceManagerIds(owner) {
		return this.#balanceManagerQueries.getBalanceManagerIds(owner);
	}
	accountExists(poolKey, managerKey) {
		return this.#balanceManagerQueries.accountExists(poolKey, managerKey);
	}
	whitelisted(poolKey) {
		return this.#poolQueries.whitelisted(poolKey);
	}
	vaultBalances(poolKey) {
		return this.#poolQueries.vaultBalances(poolKey);
	}
	getPoolIdByAssets(baseType, quoteType) {
		return this.#poolQueries.getPoolIdByAssets(baseType, quoteType);
	}
	midPrice(poolKey) {
		return this.#poolQueries.midPrice(poolKey);
	}
	poolTradeParams(poolKey) {
		return this.#poolQueries.poolTradeParams(poolKey);
	}
	poolBookParams(poolKey) {
		return this.#poolQueries.poolBookParams(poolKey);
	}
	stablePool(poolKey) {
		return this.#poolQueries.stablePool(poolKey);
	}
	registeredPool(poolKey) {
		return this.#poolQueries.registeredPool(poolKey);
	}
	poolTradeParamsNext(poolKey) {
		return this.#poolQueries.poolTradeParamsNext(poolKey);
	}
	quorum(poolKey) {
		return this.#poolQueries.quorum(poolKey);
	}
	poolId(poolKey) {
		return this.#poolQueries.poolId(poolKey);
	}
	canPlaceLimitOrder(params) {
		return this.#poolQueries.canPlaceLimitOrder(params);
	}
	canPlaceMarketOrder(params) {
		return this.#poolQueries.canPlaceMarketOrder(params);
	}
	checkMarketOrderParams(poolKey, quantity) {
		return this.#poolQueries.checkMarketOrderParams(poolKey, quantity);
	}
	checkLimitOrderParams(poolKey, price, quantity, expireTimestamp) {
		return this.#poolQueries.checkLimitOrderParams(poolKey, price, quantity, expireTimestamp);
	}
	getQuoteQuantityOut(poolKey, baseQuantity) {
		return this.#quantityQueries.getQuoteQuantityOut(poolKey, baseQuantity);
	}
	getBaseQuantityOut(poolKey, quoteQuantity) {
		return this.#quantityQueries.getBaseQuantityOut(poolKey, quoteQuantity);
	}
	getQuantityOut(poolKey, baseQuantity, quoteQuantity) {
		return this.#quantityQueries.getQuantityOut(poolKey, baseQuantity, quoteQuantity);
	}
	getQuoteQuantityOutInputFee(poolKey, baseQuantity) {
		return this.#quantityQueries.getQuoteQuantityOutInputFee(poolKey, baseQuantity);
	}
	getBaseQuantityOutInputFee(poolKey, quoteQuantity) {
		return this.#quantityQueries.getBaseQuantityOutInputFee(poolKey, quoteQuantity);
	}
	getQuantityOutInputFee(poolKey, baseQuantity, quoteQuantity) {
		return this.#quantityQueries.getQuantityOutInputFee(poolKey, baseQuantity, quoteQuantity);
	}
	getBaseQuantityIn(poolKey, targetQuoteQuantity, payWithDeep) {
		return this.#quantityQueries.getBaseQuantityIn(poolKey, targetQuoteQuantity, payWithDeep);
	}
	getQuoteQuantityIn(poolKey, targetBaseQuantity, payWithDeep) {
		return this.#quantityQueries.getQuoteQuantityIn(poolKey, targetBaseQuantity, payWithDeep);
	}
	getOrderDeepRequired(poolKey, baseQuantity, price) {
		return this.#quantityQueries.getOrderDeepRequired(poolKey, baseQuantity, price);
	}
	accountOpenOrders(poolKey, managerKey) {
		return this.#orderQueries.accountOpenOrders(poolKey, managerKey);
	}
	getOrder(poolKey, orderId) {
		return this.#orderQueries.getOrder(poolKey, orderId);
	}
	getOrderNormalized(poolKey, orderId) {
		return this.#orderQueries.getOrderNormalized(poolKey, orderId);
	}
	getOrders(poolKey, orderIds) {
		return this.#orderQueries.getOrders(poolKey, orderIds);
	}
	getLevel2Range(poolKey, priceLow, priceHigh, isBid) {
		return this.#orderQueries.getLevel2Range(poolKey, priceLow, priceHigh, isBid);
	}
	getLevel2TicksFromMid(poolKey, ticks) {
		return this.#orderQueries.getLevel2TicksFromMid(poolKey, ticks);
	}
	getAccountOrderDetails(poolKey, managerKey) {
		return this.#orderQueries.getAccountOrderDetails(poolKey, managerKey);
	}
	account(poolKey, managerKey) {
		return this.#accountQueries.account(poolKey, managerKey);
	}
	lockedBalance(poolKey, balanceManagerKey) {
		return this.#accountQueries.lockedBalance(poolKey, balanceManagerKey);
	}
	getPoolDeepPrice(poolKey) {
		return this.#accountQueries.getPoolDeepPrice(poolKey);
	}
	balanceManagerReferralOwner(referral) {
		return this.#referralQueries.balanceManagerReferralOwner(referral);
	}
	getPoolReferralBalances(poolKey, referral) {
		return this.#referralQueries.getPoolReferralBalances(poolKey, referral);
	}
	balanceManagerReferralPoolId(referral) {
		return this.#referralQueries.balanceManagerReferralPoolId(referral);
	}
	poolReferralMultiplier(poolKey, referral) {
		return this.#referralQueries.poolReferralMultiplier(poolKey, referral);
	}
	getBalanceManagerReferralId(managerKey, poolKey) {
		return this.#referralQueries.getBalanceManagerReferralId(managerKey, poolKey);
	}
	getPriceInfoObject(tx, coinKey) {
		return this.#priceFeedQueries.getPriceInfoObject(tx, coinKey);
	}
	getPriceInfoObjects(tx, coinKeys) {
		return this.#priceFeedQueries.getPriceInfoObjects(tx, coinKeys);
	}
	getPriceInfoObjectAge(coinKey) {
		return this.#priceFeedQueries.getPriceInfoObjectAge(coinKey);
	}
	getMarginPoolId(coinKey) {
		return this.#marginPoolQueries.getMarginPoolId(coinKey);
	}
	isDeepbookPoolAllowed(coinKey, deepbookPoolId) {
		return this.#marginPoolQueries.isDeepbookPoolAllowed(coinKey, deepbookPoolId);
	}
	getMarginPoolTotalSupply(coinKey, decimals = 6) {
		return this.#marginPoolQueries.getMarginPoolTotalSupply(coinKey, decimals);
	}
	getMarginPoolSupplyShares(coinKey, decimals = 6) {
		return this.#marginPoolQueries.getMarginPoolSupplyShares(coinKey, decimals);
	}
	getMarginPoolTotalBorrow(coinKey, decimals = 6) {
		return this.#marginPoolQueries.getMarginPoolTotalBorrow(coinKey, decimals);
	}
	getMarginPoolBorrowShares(coinKey, decimals = 6) {
		return this.#marginPoolQueries.getMarginPoolBorrowShares(coinKey, decimals);
	}
	getMarginPoolLastUpdateTimestamp(coinKey) {
		return this.#marginPoolQueries.getMarginPoolLastUpdateTimestamp(coinKey);
	}
	getMarginPoolSupplyCap(coinKey, decimals = 6) {
		return this.#marginPoolQueries.getMarginPoolSupplyCap(coinKey, decimals);
	}
	getMarginPoolMaxUtilizationRate(coinKey) {
		return this.#marginPoolQueries.getMarginPoolMaxUtilizationRate(coinKey);
	}
	getMarginPoolProtocolSpread(coinKey) {
		return this.#marginPoolQueries.getMarginPoolProtocolSpread(coinKey);
	}
	getMarginPoolMinBorrow(coinKey, decimals = 6) {
		return this.#marginPoolQueries.getMarginPoolMinBorrow(coinKey, decimals);
	}
	getMarginPoolInterestRate(coinKey) {
		return this.#marginPoolQueries.getMarginPoolInterestRate(coinKey);
	}
	getUserSupplyShares(coinKey, supplierCapId, decimals = 6) {
		return this.#marginPoolQueries.getUserSupplyShares(coinKey, supplierCapId, decimals);
	}
	getUserSupplyAmount(coinKey, supplierCapId, decimals = 6) {
		return this.#marginPoolQueries.getUserSupplyAmount(coinKey, supplierCapId, decimals);
	}
	getMarginManagerOwner(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerOwner(marginManagerKey);
	}
	getMarginManagerDeepbookPool(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerDeepbookPool(marginManagerKey);
	}
	getMarginManagerMarginPoolId(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerMarginPoolId(marginManagerKey);
	}
	getMarginManagerBorrowedShares(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerBorrowedShares(marginManagerKey);
	}
	getMarginManagerBorrowedBaseShares(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerBorrowedBaseShares(marginManagerKey);
	}
	getMarginManagerBorrowedQuoteShares(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerBorrowedQuoteShares(marginManagerKey);
	}
	getMarginManagerHasBaseDebt(marginManagerKey) {
		return this.#marginManagerQueries.getMarginManagerHasBaseDebt(marginManagerKey);
	}
	getMarginManagerBalanceManagerId(marginManagerAddress) {
		return this.#marginManagerQueries.getMarginManagerBalanceManagerId(marginManagerAddress);
	}
	getMarginManagerAssets(marginManagerKey, decimals = 6) {
		return this.#marginManagerQueries.getMarginManagerAssets(marginManagerKey, decimals);
	}
	getMarginManagerDebts(marginManagerKey, decimals = 6) {
		return this.#marginManagerQueries.getMarginManagerDebts(marginManagerKey, decimals);
	}
	getMarginManagerState(marginManagerKey, decimals = 6) {
		return this.#marginManagerQueries.getMarginManagerState(marginManagerKey, decimals);
	}
	getMarginManagerStates(marginManagers, decimals = 6) {
		return this.#marginManagerQueries.getMarginManagerStates(marginManagers, decimals);
	}
	getMarginManagerBaseBalance(marginManagerKey, decimals = 9) {
		return this.#marginManagerQueries.getMarginManagerBaseBalance(marginManagerKey, decimals);
	}
	getMarginManagerQuoteBalance(marginManagerKey, decimals = 9) {
		return this.#marginManagerQueries.getMarginManagerQuoteBalance(marginManagerKey, decimals);
	}
	getMarginManagerDeepBalance(marginManagerKey, decimals = 6) {
		return this.#marginManagerQueries.getMarginManagerDeepBalance(marginManagerKey, decimals);
	}
	getMarginManagerBalances(marginManagers, decimals = 9) {
		return this.#marginManagerQueries.getMarginManagerBalances(marginManagers, decimals);
	}
	getConditionalOrderIds(marginManagerKey) {
		return this.#tpslQueries.getConditionalOrderIds(marginManagerKey);
	}
	getLowestTriggerAbovePrice(marginManagerKey) {
		return this.#tpslQueries.getLowestTriggerAbovePrice(marginManagerKey);
	}
	getHighestTriggerBelowPrice(marginManagerKey) {
		return this.#tpslQueries.getHighestTriggerBelowPrice(marginManagerKey);
	}
	isPoolEnabledForMargin(poolKey) {
		return this.#registryQueries.isPoolEnabledForMargin(poolKey);
	}
	getMarginManagerIdsForOwner(owner) {
		return this.#registryQueries.getMarginManagerIdsForOwner(owner);
	}
	getBaseMarginPoolId(poolKey) {
		return this.#registryQueries.getBaseMarginPoolId(poolKey);
	}
	getQuoteMarginPoolId(poolKey) {
		return this.#registryQueries.getQuoteMarginPoolId(poolKey);
	}
	getMinWithdrawRiskRatio(poolKey) {
		return this.#registryQueries.getMinWithdrawRiskRatio(poolKey);
	}
	getMinBorrowRiskRatio(poolKey) {
		return this.#registryQueries.getMinBorrowRiskRatio(poolKey);
	}
	getLiquidationRiskRatio(poolKey) {
		return this.#registryQueries.getLiquidationRiskRatio(poolKey);
	}
	getTargetLiquidationRiskRatio(poolKey) {
		return this.#registryQueries.getTargetLiquidationRiskRatio(poolKey);
	}
	getUserLiquidationReward(poolKey) {
		return this.#registryQueries.getUserLiquidationReward(poolKey);
	}
	getPoolLiquidationReward(poolKey) {
		return this.#registryQueries.getPoolLiquidationReward(poolKey);
	}
	getAllowedMaintainers() {
		return this.#registryQueries.getAllowedMaintainers();
	}
	getAllowedPauseCaps() {
		return this.#registryQueries.getAllowedPauseCaps();
	}
	decodeOrderId(encodedOrderId) {
		return {
			isBid: encodedOrderId >> 127n === 0n,
			price: Number(encodedOrderId >> 64n & (1n << 63n) - 1n),
			orderId: Number(encodedOrderId & (1n << 64n) - 1n)
		};
	}
};

//#endregion
export { DeepBookClient, deepbook };
